{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Importing all the required modules"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import nferx\n",
    "import os\n",
    "#from urllib.parse import urlencode as uenc\n",
    "import requests\n",
    "import json  \n",
    "import urllib\n",
    "import datetime\n",
    "import configparser\n",
    "import simplejson as json\n",
    "import config\n",
    "#from config import nf\n",
    "import pandas as pd\n",
    "\n",
    "from functools import reduce\n",
    "from concurrent.futures import ThreadPoolExecutor\n",
    "import requests\n",
    "\n",
    "from urllib.parse import urlencode\n",
    "import requests\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Setting all the environment variables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "os.environ['Nfer_Access_Key'] = \"praveen@nference.net\"   ### Feel free to use this or change it to match your credentials\n",
    "os.environ['Nfer_Secret_Key'] = \"81d4782ff7cda46fdbd1c2b758af0759\"\n",
    "os.environ['Nfer_Endpoint'] = \"https://pre-staging.nferx.com/dv/201912\"  ### Hardcoding to make sure that we stick to the dataversion - 201912 as before\n",
    "rm = nferx.ResourceManager(ssl=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "PRESTAGING_COOKIE = {\n",
    "    \"csrftoken\": \"aBo9srlYDadW4hAExdYhqzawSkHQoka8\",   ### Need to make sure that this hasn't expired or changed\n",
    "    \"sessionid\": \"t0bylzf7sxsygpgqtn890ljlau52kg5t\"\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "queryExpansion_endpt = \"https://pre-staging.nferx.com/dv/201912/QueryParser/v1/parse_token\" \n",
    "def queryExpansion(query):\n",
    "    \"\"\"\n",
    "    returns the expanded query\n",
    "    \"\"\"\n",
    "    url = queryExpansion_endpt\n",
    "    qp_query = {\n",
    "        \"token\": query\n",
    "    }\n",
    "    qp_results = requests.get(url, params=qp_query, cookies=PRESTAGING_COOKIE)\n",
    "    qp_results = json.loads(qp_results.text)\n",
    "    return qp_results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def extract_syns(qe_response):\n",
    "    \"\"\"\n",
    "    Takes the output of query expansion and returns the synonym tokens\n",
    "    \"\"\"\n",
    "    ret = []\n",
    "    exp_query = qe_response['expanded_query']\n",
    "    if \";\" or \",\" in exp_query:\n",
    "     exp_query.replace(\";\",\",\")\n",
    "     ret = exp_query.split(\",\")\n",
    "    else :\n",
    "      ret = [exp_query]\n",
    "    return ret"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "#default settings; can be changed https://github.com/lumenbiomics/Entity-Extraction-Service/wiki\n",
    "qparams_dict = {\n",
    "     \"subsume\":\"type-wise\",\n",
    "     \"onlyGoodSyn\":1,\n",
    "     \"sendData\":1,\n",
    "     \"eesMode\":2,\n",
    "     \"onlyGoodSyn\":1,\n",
    "     \"needWordWindow\":1,\n",
    "     \"wordWindow\":15,\n",
    "    \"showAkasInfo\":1\n",
    "    #\"whitelistTypes\":\"disease,drug\"\n",
    "   }\n",
    "headers_dict = {\n",
    "   \"Accept\": \"*/*\" ,\n",
    "   \"Accept-Encoding\": \"gzip, deflate\", \n",
    "   \"Content-Type\": \"application/json\"\n",
    "   }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "#to use this function you need to pass \n",
    "# ees_endpt: where is the EES running\n",
    "# textchunks_list: list of text strings that you want to extract entities from \n",
    "#the other two args are defined above and you can reuse them \n",
    "#returns a json string ; deconstructed below\n",
    "\n",
    "def ees_get_entities(ees_endpt, qparams_dict, textchunks_list, headers_dict):\n",
    "\n",
    "    url = ees_endpt+\"/EntityExtraction/v3/getentities?\"\n",
    "    url+=urllib.parse.urlencode(qparams_dict)\n",
    "    jsontext = [ {\"text\": e} for e in textchunks_list]\n",
    "    r = requests.post(url,data=json.dumps(jsontext),headers=headers_dict,verify=False)\n",
    "    return r \n",
    "\n",
    "#eed endpt\n",
    "ees_endpt = \"http://n1-r20.w2v.servers.nferx.com\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Just testing to see if get entities is working\n",
    "textchunks_list = [\n",
    "         'naproxen ,  ketoprofen ,  flurbiprofen ,  xaprozin ,  phenylbutazone ,  fenbufen ,  loxoprofen_sodium ,  pranoprofen ,  aurothiomalate ,  sodium_aurothiomalate ,  morphine ,  zaltoprofen ,   probenecid ,  sulfinpyrazone ,  colchicine ,  allopurinol ,  benzbromarone ,  tiaramide , hyaluronate ,  camostat ,  morphine_hydrochloride,  colchicine_probenecid ,  ulinastatin , gabexate_mesilate'\n",
    "]\n",
    "#call to ees with the text chunks - this will give q secureRequest warning for now \n",
    "r=ees_get_entities(ees_endpt, qparams_dict, textchunks_list, headers_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_Entities(listofsignals):\n",
    "    print(\"Entering get_Entities\",datetime.datetime.now().time())\n",
    "    estr = \" \".join(listofsignals)\n",
    "    textchunks_list = [estr]\n",
    "    r=ees_get_entities(ees_endpt, qparams_dict, textchunks_list, headers_dict)\n",
    "    jresp = json.loads(r.content)\n",
    "    retlist = []\n",
    "    origPhrase = []\n",
    "    for e in jresp[\"result\"][0]['entities']:\n",
    "        if e['score']>=1:\n",
    "            retlist.append(e['display_name'])\n",
    "            origPhrase.append(e['original_phrase'])\n",
    "    print(\"Leaving get_Entities\",datetime.datetime.now().time())\n",
    "    return retlist,origPhrase"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "def getSignals(query, corpus, collections=\"root-drugs,root-genes,root-diseases,root-biomolecules\",expandQuery=True):\n",
    "    inst.Query=query\n",
    "    inst.Page=0\n",
    "    inst.Limit=0\n",
    "    inst.Corpus = corpus\n",
    "    inst.SmartQuery = expandQuery\n",
    "    #inst.MaxFragments=100000\n",
    "    #inst.WindowWidth=50\n",
    "    inst.Collections=collections\n",
    "    #x=inst.get_relevant_patient_docs()\n",
    "    x=inst.get_signals()\n",
    "    #print(inst.Query)\n",
    "    return x"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'original_token': 'bermekimab',\n",
       " 'spell_corrected': False,\n",
       " 'success': True,\n",
       " 'modified_query': 'bermekimab',\n",
       " 'expanded_query': 'bermekimab,mabp1,xilonix,t2_18c3,ra_18c3,cv_18c3,ca_18c3,bermekimabs,ca18c3,cv18c3,ra18c3,t218c3',\n",
       " 'display_query': 'bermekimab,mabp1,xilonix,t2_18c3,ra_18c3,cv_18c3,ca_18c3,bermekimabs,ca18c3,cv18c3,ra18c3,t218c3',\n",
       " 'did_you_mean': None,\n",
       " 'suggestions': {},\n",
       " 'result': 'bermekimab,mabp1,xilonix,t2_18c3,ra_18c3,cv_18c3,ca_18c3,bermekimabs,ca18c3,cv18c3,ra18c3,t218c3',\n",
       " 'detailed': [[[{'entity_id': 97561,\n",
       "     'token': 'bermekimab',\n",
       "     'type': 'drug',\n",
       "     'syns': ['mabp1',\n",
       "      'xilonix',\n",
       "      'bermekimab',\n",
       "      't2_18c3',\n",
       "      'ra_18c3',\n",
       "      'cv_18c3',\n",
       "      'ca_18c3',\n",
       "      'bermekimabs',\n",
       "      'ca18c3',\n",
       "      'cv18c3',\n",
       "      'ra18c3',\n",
       "      't218c3'],\n",
       "     'variants': ['bermekimab'],\n",
       "     'suggested_entities': [],\n",
       "     'entity_expansions': None,\n",
       "     'spell_corrections': None,\n",
       "     'kg_info': {'id': 'ROOT/7bd3f88d-2bca-44cc-8bb2-a161d0ab4b09',\n",
       "      'data_version': 'nucleus_v1_4_1'}}]]]}"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bermekimab_syn_query = queryExpansion(\"bermekimab\")\n",
    "bermekimab_syn_query"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'original_token': 'IL-1A',\n",
       " 'spell_corrected': False,\n",
       " 'success': True,\n",
       " 'modified_query': 'IL-1A',\n",
       " 'expanded_query': 'il-1a,il_1α,il1,il_1_alpha,interleukin_1_alpha,interleukin_1α,il_1_α,il1_α,il1_alpha,il1alpha,interleukin_1_α,il1a gene,hematopoietin_1,il1f1,interleukins_1_α,interleukins_1_alpha,interleukins_1α,pro_interleukin_1_alpha,preinterleukin_1_alpha,receptor interleukin_1,il1a genes,il_1a,il 1a,il1_a,il1a',\n",
       " 'display_query': 'il-1a,il_1α,il1,il_1_alpha,interleukin_1_alpha,interleukin_1α,il_1_α,il1_α,il1_alpha,il1alpha,interleukin_1_α,il1a gene,hematopoietin_1,il1f1,interleukins_1_α,interleukins_1_alpha,interleukins_1α,pro_interleukin_1_alpha,preinterleukin_1_alpha,receptor interleukin_1,il1a genes',\n",
       " 'did_you_mean': None,\n",
       " 'suggestions': {},\n",
       " 'result': 'il-1a,il_1α,il1,il_1_alpha,interleukin_1_alpha,interleukin_1α,il_1_α,il1_α,il1_alpha,il1alpha,interleukin_1_α,il1a gene,hematopoietin_1,il1f1,interleukins_1_α,interleukins_1_alpha,interleukins_1α,pro_interleukin_1_alpha,preinterleukin_1_alpha,receptor interleukin_1,il1a genes,il_1a,il 1a,il1_a,il1a',\n",
       " 'detailed': [[[{'entity_id': 18019,\n",
       "     'token': 'il-1a',\n",
       "     'type': 'gene',\n",
       "     'syns': ['il_1α',\n",
       "      'il1',\n",
       "      'il_1_alpha',\n",
       "      'interleukin_1_alpha',\n",
       "      'interleukin_1α',\n",
       "      'il_1_α',\n",
       "      'il1_α',\n",
       "      'il1_alpha',\n",
       "      'il1alpha',\n",
       "      'interleukin_1_α',\n",
       "      'il1a gene',\n",
       "      'hematopoietin_1',\n",
       "      'il1f1',\n",
       "      'interleukins_1_α',\n",
       "      'interleukins_1_alpha',\n",
       "      'interleukins_1α',\n",
       "      'pro_interleukin_1_alpha',\n",
       "      'preinterleukin_1_alpha',\n",
       "      'receptor interleukin_1',\n",
       "      'il1a genes'],\n",
       "     'variants': ['il_1a', 'il 1a'],\n",
       "     'suggested_entities': [],\n",
       "     'entity_expansions': None,\n",
       "     'spell_corrections': None,\n",
       "     'kg_info': {'id': 'ROOT/473c41f3-840d-4949-a70a-3c8b525d1898',\n",
       "      'data_version': 'nucleus_v1_4_1'}}]]]}"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "il1a_syn_query = queryExpansion(\"IL-1A\")\n",
    "il1a_syn_query"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'original_token': 'IL-1B',\n",
       " 'spell_corrected': False,\n",
       " 'success': True,\n",
       " 'modified_query': 'IL-1B',\n",
       " 'expanded_query': 'il-1b,il_1β,il_1,interleukin_1β,il_1_β,il_1_beta,interleukin_1_beta,il1_β,interleukin_1_β,il1_beta,il1beta,il1b gene,catabolin,interleukins_1β,il1f2,il_1ss,il1b genes,pro_interleukin_1_beta,interleukins_1_beta,interleukins_1_β,preinterleukin_1_beta,il_1b,il 1b,il1_b,il1b',\n",
       " 'display_query': 'il-1b,il_1β,il_1,interleukin_1β,il_1_β,il_1_beta,interleukin_1_beta,il1_β,interleukin_1_β,il1_beta,il1beta,il1b gene,catabolin,interleukins_1β,il1f2,il_1ss,il1b genes,pro_interleukin_1_beta,interleukins_1_beta,interleukins_1_β,preinterleukin_1_beta',\n",
       " 'did_you_mean': None,\n",
       " 'suggestions': {},\n",
       " 'result': 'il-1b,il_1β,il_1,interleukin_1β,il_1_β,il_1_beta,interleukin_1_beta,il1_β,interleukin_1_β,il1_beta,il1beta,il1b gene,catabolin,interleukins_1β,il1f2,il_1ss,il1b genes,pro_interleukin_1_beta,interleukins_1_beta,interleukins_1_β,preinterleukin_1_beta,il_1b,il 1b,il1_b,il1b',\n",
       " 'detailed': [[[{'entity_id': 18020,\n",
       "     'token': 'il-1b',\n",
       "     'type': 'gene',\n",
       "     'syns': ['il_1β',\n",
       "      'il_1',\n",
       "      'interleukin_1β',\n",
       "      'il_1_β',\n",
       "      'il_1_beta',\n",
       "      'interleukin_1_beta',\n",
       "      'il1_β',\n",
       "      'interleukin_1_β',\n",
       "      'il1_beta',\n",
       "      'il1beta',\n",
       "      'il1b gene',\n",
       "      'catabolin',\n",
       "      'interleukins_1β',\n",
       "      'il1f2',\n",
       "      'il_1ss',\n",
       "      'il1b genes',\n",
       "      'pro_interleukin_1_beta',\n",
       "      'interleukins_1_beta',\n",
       "      'interleukins_1_β',\n",
       "      'preinterleukin_1_beta'],\n",
       "     'variants': ['il_1b', 'il 1b'],\n",
       "     'suggested_entities': [],\n",
       "     'entity_expansions': None,\n",
       "     'spell_corrections': None,\n",
       "     'kg_info': {'id': 'ROOT/9f5f0966-7e69-4518-9051-e25d99451ba3',\n",
       "      'data_version': 'nucleus_v1_4_1'}}]]]}"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "il1b_syn_query = queryExpansion(\"IL-1B\")\n",
    "il1b_syn_query"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'original_token': 'IL1R1',\n",
       " 'spell_corrected': False,\n",
       " 'success': True,\n",
       " 'modified_query': 'IL1R1',\n",
       " 'expanded_query': 'il1r1,p80,il1ra,il1r,cd121a,interleukin_1_receptor type_1,interleukin_1_receptor alpha,il_1rt1,il1rs,il_1r_alpha,il_1r_α,d2s1473,cd121a antigen,il_1rt_1,interleukin_1_receptor_α,interleukin_1_receptor type_i,cd121 antigen like family member_a,il1ras,interleukin_1_receptor alpha type_i,il1_r_1,il1 r1,il1_r1',\n",
       " 'display_query': 'il1r1,p80,il1ra,il1r,cd121a,interleukin_1_receptor type_1,interleukin_1_receptor alpha,il_1rt1,il1rs,il_1r_alpha,il_1r_α,d2s1473,cd121a antigen,il_1rt_1,interleukin_1_receptor_α,interleukin_1_receptor type_i,cd121 antigen like family member_a,il1ras,interleukin_1_receptor alpha type_i',\n",
       " 'did_you_mean': None,\n",
       " 'suggestions': {},\n",
       " 'result': 'il1r1,p80,il1ra,il1r,cd121a,interleukin_1_receptor type_1,interleukin_1_receptor alpha,il_1rt1,il1rs,il_1r_alpha,il_1r_α,d2s1473,cd121a antigen,il_1rt_1,interleukin_1_receptor_α,interleukin_1_receptor type_i,cd121 antigen like family member_a,il1ras,interleukin_1_receptor alpha type_i,il1_r_1,il1 r1,il1_r1',\n",
       " 'detailed': [[[{'entity_id': 498287,\n",
       "     'token': 'il1r1',\n",
       "     'type': 'receptor',\n",
       "     'syns': ['p80',\n",
       "      'il1ra',\n",
       "      'il1r',\n",
       "      'cd121a',\n",
       "      'interleukin_1_receptor type_1',\n",
       "      'interleukin_1_receptor alpha',\n",
       "      'il_1rt1',\n",
       "      'il1rs',\n",
       "      'il_1r_alpha',\n",
       "      'il_1r_α',\n",
       "      'd2s1473',\n",
       "      'cd121a antigen',\n",
       "      'il_1rt_1',\n",
       "      'interleukin_1_receptor_α',\n",
       "      'interleukin_1_receptor type_i',\n",
       "      'cd121 antigen like family member_a',\n",
       "      'il1ras',\n",
       "      'interleukin_1_receptor alpha type_i'],\n",
       "     'variants': ['il1r1'],\n",
       "     'suggested_entities': [],\n",
       "     'entity_expansions': None,\n",
       "     'spell_corrections': None,\n",
       "     'kg_info': {'id': 'ROOT/6540b5a6-e2f5-401a-a9f4-283ea072f0ff',\n",
       "      'data_version': 'nucleus_v1_4_1'}}]]]}"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "il1r1_syn_query = queryExpansion(\"IL1R1\")\n",
    "il1r1_syn_query"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'original_token': 'IL1RN',\n",
       " 'spell_corrected': False,\n",
       " 'success': True,\n",
       " 'modified_query': 'IL1RN',\n",
       " 'expanded_query': 'il1rn,interleukin_1_receptor antagonist,il_1_receptor antagonist,il_1_antagonist,il_1_receptor antagonists,il_1_antagonists,icil_1ra,interleukin_1_receptor antagonists,interleukin_1_receptor antagonist protein,il1rn gene,interleukin_1_receptor antagonistic protein,il1f3,interleukin receptor antagonist,il_1ra3,il1rn genes,il1_inhibitor,mvcd4,mgc10430,icil1ra,il1_inhibitors,intracellular interleukin_1_receptor antagonist',\n",
       " 'display_query': 'il1rn,interleukin_1_receptor antagonist,il_1_receptor antagonist,il_1_antagonist,il_1_receptor antagonists,il_1_antagonists,icil_1ra,interleukin_1_receptor antagonists,interleukin_1_receptor antagonist protein,il1rn gene,interleukin_1_receptor antagonistic protein,il1f3,interleukin receptor antagonist,il_1ra3,il1rn genes,il1_inhibitor,mvcd4,mgc10430,icil1ra,il1_inhibitors,intracellular interleukin_1_receptor antagonist',\n",
       " 'did_you_mean': None,\n",
       " 'suggestions': {},\n",
       " 'result': 'il1rn,interleukin_1_receptor antagonist,il_1_receptor antagonist,il_1_antagonist,il_1_receptor antagonists,il_1_antagonists,icil_1ra,interleukin_1_receptor antagonists,interleukin_1_receptor antagonist protein,il1rn gene,interleukin_1_receptor antagonistic protein,il1f3,interleukin receptor antagonist,il_1ra3,il1rn genes,il1_inhibitor,mvcd4,mgc10430,icil1ra,il1_inhibitors,intracellular interleukin_1_receptor antagonist',\n",
       " 'detailed': [[[{'entity_id': 18030,\n",
       "     'token': 'il1rn',\n",
       "     'type': 'gene',\n",
       "     'syns': ['interleukin_1_receptor antagonist',\n",
       "      'il_1_receptor antagonist',\n",
       "      'il_1_antagonist',\n",
       "      'il_1_receptor antagonists',\n",
       "      'il_1_antagonists',\n",
       "      'icil_1ra',\n",
       "      'interleukin_1_receptor antagonists',\n",
       "      'interleukin_1_receptor antagonist protein',\n",
       "      'il1rn gene',\n",
       "      'interleukin_1_receptor antagonistic protein',\n",
       "      'il1f3',\n",
       "      'interleukin receptor antagonist',\n",
       "      'il_1ra3',\n",
       "      'il1rn genes',\n",
       "      'il1_inhibitor',\n",
       "      'mvcd4',\n",
       "      'mgc10430',\n",
       "      'icil1ra',\n",
       "      'il1_inhibitors',\n",
       "      'intracellular interleukin_1_receptor antagonist'],\n",
       "     'variants': ['il1rn'],\n",
       "     'suggested_entities': [],\n",
       "     'entity_expansions': None,\n",
       "     'spell_corrections': None,\n",
       "     'kg_info': {'id': 'ROOT/41bef8d2-86bf-487e-8855-2391344f6938',\n",
       "      'data_version': 'nucleus_v1_4_1'}}]]]}"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "il1ra_syn_query = queryExpansion(\"IL1RN\")\n",
    "il1ra_syn_query"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Final list of synonyms chosen after discussion with Janssen team"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "synonyms = {'bermekimab':['mabp1',\n",
    "      'xilonix',\n",
    "      'bermekimab', ## Making sure the original query token is present\n",
    "      't2_18c3',\n",
    "      'ra_18c3',\n",
    "      'cv_18c3',\n",
    "      'ca_18c3',\n",
    "      'bermekimabs',\n",
    "      'ca18c3',\n",
    "      'cv18c3',\n",
    "      'ra18c3',\n",
    "      't218c3'], \n",
    "               'IL-1A': ['IL-1A', ## Making sure the original query token is present\n",
    "                   'il_1α',\n",
    "      'il1',\n",
    "      'il_1_alpha',\n",
    "      'interleukin_1_alpha',\n",
    "      'interleukin_1α',\n",
    "      'il_1_α',\n",
    "      'il1_α',\n",
    "      'il1_alpha',\n",
    "      'il1alpha',\n",
    "      'interleukin_1_α',\n",
    "      'il1a gene',\n",
    "      'hematopoietin_1',\n",
    "      'il1f1',\n",
    "      'interleukins_1_α',\n",
    "      'interleukins_1_alpha',\n",
    "      'interleukins_1α',\n",
    "      'pro_interleukin_1_alpha',\n",
    "      'preinterleukin_1_alpha',\n",
    "      'receptor interleukin_1',\n",
    "      'il1a genes'],\n",
    "            \n",
    "            'IL-1B': ['IL-1B', ## Making sure the original query token is present\n",
    "                   'il_1β',\n",
    "      'il_1',\n",
    "      'interleukin_1β',\n",
    "      'il_1_β',\n",
    "      'il_1_beta',\n",
    "      'interleukin_1_beta',\n",
    "      'il1_β',\n",
    "      'interleukin_1_β',\n",
    "      'il1_beta',\n",
    "      'il1beta',\n",
    "      'il1b gene',\n",
    "      'catabolin',\n",
    "      'interleukins_1β',\n",
    "      'il1f2',\n",
    "      'il_1ss',\n",
    "      'il1b genes',\n",
    "      'pro_interleukin_1_beta',\n",
    "      'interleukins_1_beta',\n",
    "      'interleukins_1_β',\n",
    "      'preinterleukin_1_beta'],\n",
    "            \n",
    "            'IL-1R': [ 'IL1R1', ## Making sure the original query token is present\n",
    "                'p80',\n",
    "      'il1ra',\n",
    "      'il1r',\n",
    "      'cd121a',\n",
    "      'interleukin_1_receptor type_1',\n",
    "      'interleukin_1_receptor alpha',\n",
    "      'il_1rt1',\n",
    "      'il1rs',\n",
    "      'il_1r_alpha',\n",
    "      'il_1r_α',\n",
    "      'd2s1473',\n",
    "      'cd121a antigen',\n",
    "      'il_1rt_1',\n",
    "      'interleukin_1_receptor_α',\n",
    "      'interleukin_1_receptor type_i',\n",
    "      'cd121 antigen like family member_a',\n",
    "      'il1ras',\n",
    "      'interleukin_1_receptor alpha type_i'],\n",
    "            \n",
    "            \n",
    "                'IL-1RA': [ 'IL1RN', ## Making sure the original query token is present\n",
    "       'interleukin_1_receptor antagonist',\n",
    "      'il_1_receptor antagonist',\n",
    "      'il_1_antagonist',\n",
    "      'il_1_receptor antagonists',\n",
    "      'il_1_antagonists',\n",
    "      'icil_1ra',\n",
    "      'interleukin_1_receptor antagonists',\n",
    "      'interleukin_1_receptor antagonist protein',\n",
    "      'il1rn gene',\n",
    "      'interleukin_1_receptor antagonistic protein',\n",
    "      'il1f3',\n",
    "      'interleukin receptor antagonist',\n",
    "      'il_1ra3',\n",
    "      'il1rn genes',\n",
    "      'il1_inhibitor',\n",
    "      'mvcd4',\n",
    "      'mgc10430',\n",
    "      'icil1ra',\n",
    "      #'il1_inhibitors', ### Commenting this out as it is non-specific\n",
    "      'intracellular interleukin_1_receptor antagonist']\n",
    "           }"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Using computesignals url to get the results of global and local score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "url = 'https://pre-staging.nferx.com/dv/201912/semantics/v2/compute_signals?'\n",
    "\n",
    "\n",
    "###Using Dee's multithreader wrapped function to perform the query and extract out the signals\n",
    "exe = ThreadPoolExecutor(max_workers=5)\n",
    "\n",
    "params = {'window': 50, \n",
    "          'sort_by': 'local', \n",
    "         'corpus': 'corpus', \n",
    "          'max_count': 'infinity',\n",
    "          'synonyms': True, \n",
    "          #'token':'IL-1A', \n",
    "          'control':'root-diseases',\n",
    "         }\n",
    "\n",
    "def query_lab(query, collection, window = 101, subset = ['score',  'sig', 'token'], synonyms = True, cookies=PRESTAGING_COOKIE):\n",
    "    \n",
    "    print('Fetching : %s'%query)\n",
    "    params['token'] = query\n",
    "    params['control'] = collection\n",
    "    params['synonyms'] = synonyms\n",
    "    #headers = nf.get_req_headers()\n",
    "    res = requests.get(url + urlencode(params), cookies=cookies).json()['result']['signals'][0]['signals']\n",
    "\n",
    "    try:\n",
    "        \n",
    "        if type(res)!= str:\n",
    "            res = pd.DataFrame(res)\n",
    "            res = res.loc[:, subset]\n",
    "            res.columns = (res.columns[0:-1] + '_%s'%query).to_list() + ['token']\n",
    "            res.set_index('token', inplace = True)\n",
    "    #         res['query'] = query\n",
    "    except Exception as e:\n",
    "        print('\\n%s returned no tokens'%query)\n",
    "        res = pd.DataFrame()\n",
    "        \n",
    "    #res.token = res.token.str.upper()\n",
    "    \n",
    "    return res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "def merge_non_empty(left, right):\n",
    "    if right.empty:\n",
    "        return left\n",
    "    else:\n",
    "        return left.join(right, how = 'outer')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'query_lab' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-1-c686b90ae1e1>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0mquery_terms_\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m'bermekimab'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'IL-1A'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'IL-1B'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'IL-1R'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'IL-1RA'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      2\u001b[0m signals_key_term = [query_lab(x, collection='gender', synonyms=True, subset = ['score', 'token'])\n\u001b[0;32m----> 3\u001b[0;31m                     for x in query_terms_] ### Performing query for local scores here\n\u001b[0m",
      "\u001b[0;32m<ipython-input-1-c686b90ae1e1>\u001b[0m in \u001b[0;36m<listcomp>\u001b[0;34m(.0)\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0mquery_terms_\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m'bermekimab'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'IL-1A'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'IL-1B'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'IL-1R'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'IL-1RA'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      2\u001b[0m signals_key_term = [query_lab(x, collection='gender', synonyms=True, subset = ['score', 'token'])\n\u001b[0;32m----> 3\u001b[0;31m                     for x in query_terms_] ### Performing query for local scores here\n\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m: name 'query_lab' is not defined"
     ]
    }
   ],
   "source": [
    "query_terms_ = ['bermekimab', 'IL-1A', 'IL-1B', 'IL-1R', 'IL-1RA']\n",
    "signals_key_term = [query_lab(x, collection='root-diseases', synonyms=True, subset = ['score', 'token'])\n",
    "                    for x in query_terms_] ### Performing query for local scores here"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_local_merged = reduce(lambda left, right: merge_non_empty(left, right), signals_key_term)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>score_bermekimab</th>\n",
       "      <th>score_IL-1A</th>\n",
       "      <th>score_IL-1B</th>\n",
       "      <th>score_IL-1R</th>\n",
       "      <th>score_IL-1RA</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>token</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>16p11_2_deletion_syndrome</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>17_alpha_hydroxylase_deficiency</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.029613</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>17q21_31_microdeletion_syndrome</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1p36_deletion_syndrome</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.070510</td>\n",
       "      <td>0.032861</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>21_hydroxylase_deficiency</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.036975</td>\n",
       "      <td>0.018947</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.078690</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 score_bermekimab  score_IL-1A  score_IL-1B  \\\n",
       "token                                                                         \n",
       "16p11_2_deletion_syndrome                     NaN     0.000000     0.000000   \n",
       "17_alpha_hydroxylase_deficiency               NaN          NaN          NaN   \n",
       "17q21_31_microdeletion_syndrome               NaN          NaN     0.000000   \n",
       "1p36_deletion_syndrome                        NaN     0.070510     0.032861   \n",
       "21_hydroxylase_deficiency                     NaN     0.036975     0.018947   \n",
       "\n",
       "                                 score_IL-1R  score_IL-1RA  \n",
       "token                                                       \n",
       "16p11_2_deletion_syndrome                NaN           NaN  \n",
       "17_alpha_hydroxylase_deficiency          NaN      0.029613  \n",
       "17q21_31_microdeletion_syndrome          NaN           NaN  \n",
       "1p36_deletion_syndrome                   NaN           NaN  \n",
       "21_hydroxylase_deficiency                NaN      0.078690  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_local_merged.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_local_merged.to_csv('./Final_combined_localscore.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fetching : mabp1\n",
      "Fetching : xilonixFetching : bermekimab\n",
      "\n",
      "Fetching : t2_18c3\n",
      "Fetching : ra_18c3\n",
      "Fetching : cv_18c3\n",
      "Fetching : ca_18c3Fetching : bermekimabs\n",
      "\n",
      "Fetching : ca18c3\n",
      "Fetching : cv18c3\n",
      "Fetching : ra18c3\n",
      "Fetching : t218c3\n",
      "\n",
      "ca18c3 returned no tokens\n",
      "\n",
      "cv18c3 returned no tokens\n",
      "\n",
      "ra18c3 returned no tokens\n",
      "\n",
      "t218c3 returned no tokens\n",
      "Fetching : IL-1AFetching : il_1α\n",
      "Fetching : il1Fetching : il_1_alpha\n",
      "\n",
      "Fetching : interleukin_1_alpha\n",
      "\n",
      "Fetching : interleukin_1α\n",
      "Fetching : il_1_α\n",
      "Fetching : il1_α\n",
      "Fetching : il1_alpha\n",
      "Fetching : il1alpha\n",
      "Fetching : interleukin_1_α\n",
      "Fetching : il1a gene\n",
      "Fetching : hematopoietin_1\n",
      "Fetching : il1f1\n",
      "Fetching : interleukins_1_α\n",
      "Fetching : interleukins_1_alpha\n",
      "Fetching : interleukins_1α\n",
      "Fetching : pro_interleukin_1_alpha\n",
      "Fetching : preinterleukin_1_alpha\n",
      "Fetching : receptor interleukin_1\n",
      "Fetching : il1a genes\n",
      "Fetching : IL-1B\n",
      "Fetching : il_1β\n",
      "Fetching : il_1Fetching : interleukin_1β\n",
      "Fetching : il_1_β\n",
      "\n",
      "Fetching : il_1_beta\n",
      "Fetching : interleukin_1_beta\n",
      "Fetching : il1_β\n",
      "Fetching : interleukin_1_β\n",
      "Fetching : il1_beta\n",
      "Fetching : il1beta\n",
      "Fetching : il1b gene\n",
      "Fetching : catabolin\n",
      "Fetching : interleukins_1β\n",
      "Fetching : il1f2\n",
      "Fetching : il_1ss\n",
      "Fetching : il1b genes\n",
      "Fetching : pro_interleukin_1_beta\n",
      "Fetching : interleukins_1_beta\n",
      "Fetching : interleukins_1_β\n",
      "Fetching : preinterleukin_1_beta\n",
      "Fetching : IL1R1Fetching : p80\n",
      "\n",
      "Fetching : il1ra\n",
      "Fetching : il1rFetching : cd121a\n",
      "\n",
      "Fetching : interleukin_1_receptor type_1\n",
      "Fetching : interleukin_1_receptor alpha\n",
      "Fetching : il_1rt1\n",
      "Fetching : il1rs\n",
      "Fetching : il_1r_alpha\n",
      "Fetching : il_1r_α\n",
      "Fetching : d2s1473\n",
      "Fetching : cd121a antigen\n",
      "Fetching : il_1rt_1\n",
      "Fetching : interleukin_1_receptor_α\n",
      "Fetching : interleukin_1_receptor type_i\n",
      "Fetching : cd121 antigen like family member_a\n",
      "Fetching : il1ras\n",
      "Fetching : interleukin_1_receptor alpha type_i\n",
      "\n",
      "interleukin_1_receptor alpha type_i returned no tokens\n",
      "Fetching : IL1RNFetching : interleukin_1_receptor antagonist\n",
      "Fetching : il_1_receptor antagonist\n",
      "\n",
      "Fetching : il_1_antagonist\n",
      "Fetching : il_1_receptor antagonists\n",
      "Fetching : il_1_antagonists\n",
      "Fetching : icil_1ra\n",
      "Fetching : interleukin_1_receptor antagonists\n",
      "Fetching : interleukin_1_receptor antagonist protein\n",
      "Fetching : il1rn gene\n",
      "Fetching : interleukin_1_receptor antagonistic protein\n",
      "Fetching : il1f3\n",
      "Fetching : interleukin receptor antagonist\n",
      "Fetching : il_1ra3\n",
      "Fetching : il1rn genes\n",
      "Fetching : il1_inhibitor\n",
      "Fetching : mvcd4\n",
      "Fetching : mgc10430\n",
      "Fetching : icil1ra\n",
      "Fetching : intracellular interleukin_1_receptor antagonist\n"
     ]
    }
   ],
   "source": [
    "query_terms = [term for list_ in synonyms.values() for term in list_ ]\n",
    "signals_syn_terms = {key : list(exe.map(lambda x: query_lab(x, \n",
    "                                                            collection = 'root-diseases', \n",
    "                                                            synonyms = False,\n",
    "                                                           subset = ['sig', 'token']), value))\n",
    "                     for key, value in synonyms.items()}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_global_merged = {key: reduce(lambda left, right: merge_non_empty(left, right), value)\n",
    "             for key, value in signals_syn_terms.items()}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "only_bermekimab_global_modules = df_global_merged['bermekimab']\n",
    "only_bermekimab_global_modules['maxtoken'] = list(df_global_merged['bermekimab'].idxmax(axis=1))\n",
    "only_bermekimab_global_modules.fillna(0)\n",
    "only_bermekimab_global_modules.to_csv('./Final_bermekimab_individual_token_globalscores.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys(['bermekimab', 'IL-1A', 'IL-1B', 'IL-1R', 'IL-1RA'])"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_global_merged.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "only_il1a_global_modules = df_global_merged['IL-1A']\n",
    "only_il1a_global_modules['maxtoken'] = list(df_global_merged['IL-1A'].idxmax(axis=1))\n",
    "only_il1a_global_modules.fillna(0)\n",
    "only_il1a_global_modules.to_csv('./Final_IL1A_individual_token_globalscores.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "only_il1b_global_modules = df_global_merged['IL-1B']\n",
    "only_il1b_global_modules['maxtoken'] = list(df_global_merged['IL-1B'].idxmax(axis=1))\n",
    "only_il1b_global_modules.fillna(0)\n",
    "only_il1b_global_modules.to_csv('./Final_IL1B_individual_token_globalscores.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "only_il1r_global_modules = df_global_merged['IL-1R']\n",
    "only_il1r_global_modules['maxtoken'] = list(df_global_merged['IL-1R'].idxmax(axis=1))\n",
    "only_il1r_global_modules.fillna(0)\n",
    "only_il1r_global_modules.to_csv('./Final_IL1R_individual_token_globalscores.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "only_il1ra_global_modules = df_global_merged['IL-1RA']\n",
    "only_il1ra_global_modules['maxtoken'] = list(df_global_merged['IL-1RA'].idxmax(axis=1))\n",
    "only_il1ra_global_modules.fillna(0)\n",
    "only_il1ra_global_modules.to_csv('./Final_IL1RA_individual_token_globalscores.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Storing the max of the global scores across each modules"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "#take max along cols for each disease token\n",
    "df_max_global_merged = [df.max(1).to_frame(name = 'sig_%s'%key) for key, df in df_global_merged.items()]\n",
    "df_max_global_merged = reduce(lambda left, right: merge_non_empty(left, right), df_max_global_merged)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sig_bermekimab</th>\n",
       "      <th>sig_IL-1A</th>\n",
       "      <th>sig_IL-1B</th>\n",
       "      <th>sig_IL-1R</th>\n",
       "      <th>sig_IL-1RA</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>token</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>11q_deletion_syndrome</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.82</td>\n",
       "      <td>1.58</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>11q_syndrome</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.34</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>13q_syndrome</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>15q24_deletion</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.40</td>\n",
       "      <td>1.42</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.97</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>15q24_microdeletion</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.49</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.58</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       sig_bermekimab  sig_IL-1A  sig_IL-1B  sig_IL-1R  \\\n",
       "token                                                                    \n",
       "11q_deletion_syndrome             NaN       1.82       1.58        NaN   \n",
       "11q_syndrome                      NaN       1.34        NaN        NaN   \n",
       "13q_syndrome                      NaN        NaN        NaN        NaN   \n",
       "15q24_deletion                    NaN       1.40       1.42        NaN   \n",
       "15q24_microdeletion               NaN       1.49        NaN        NaN   \n",
       "\n",
       "                       sig_IL-1RA  \n",
       "token                              \n",
       "11q_deletion_syndrome        2.04  \n",
       "11q_syndrome                  NaN  \n",
       "13q_syndrome                 0.00  \n",
       "15q24_deletion               1.97  \n",
       "15q24_microdeletion          1.58  "
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_max_global_merged.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Dees function to aggregate the columns\n",
    "\n",
    "def agg_cols(df, cutoff_sig):\n",
    "    \n",
    "    \n",
    "    #filter rows with no score > 0\n",
    "    idx_filter = (df.fillna(0) > 0).any(1)\n",
    "    df = df[idx_filter].fillna(0)\n",
    "    \n",
    "    #new agg columns\n",
    "    df['n_sig'] = (df >= cutoff_sig).sum(1)\n",
    "#     df['n_gt0'] = (df.loc[:, score_cols] > 0).sum(1)\n",
    "    df = df[df.n_sig > 0]\n",
    "    df['max_score'] = df.max(1)\n",
    "    \n",
    "    #sort values\n",
    "    df = df.sort_values(by = 'n_sig').groupby('n_sig').apply(lambda x: \n",
    "                                    x.sort_values('max_score', \n",
    "                                                  ascending = False)).reset_index(level = 0, drop = True)\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sig_bermekimab</th>\n",
       "      <th>sig_IL-1A</th>\n",
       "      <th>sig_IL-1B</th>\n",
       "      <th>sig_IL-1R</th>\n",
       "      <th>sig_IL-1RA</th>\n",
       "      <th>n_sig</th>\n",
       "      <th>max_score</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>token</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>relapsed_renal_cell_carcinoma</td>\n",
       "      <td>4.58</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>1</td>\n",
       "      <td>4.58</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>adca3</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.99</td>\n",
       "      <td>1</td>\n",
       "      <td>3.99</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>rp60</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.9</td>\n",
       "      <td>0.00</td>\n",
       "      <td>1</td>\n",
       "      <td>3.90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>recurrent_endometrial_cancer</td>\n",
       "      <td>3.82</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>1</td>\n",
       "      <td>3.82</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>metastatic_pancreatic_adenocarcinoma</td>\n",
       "      <td>3.75</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>1</td>\n",
       "      <td>3.75</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                      sig_bermekimab  sig_IL-1A  sig_IL-1B  \\\n",
       "token                                                                        \n",
       "relapsed_renal_cell_carcinoma                   4.58        0.0        0.0   \n",
       "adca3                                           0.00        0.0        0.0   \n",
       "rp60                                            0.00        0.0        0.0   \n",
       "recurrent_endometrial_cancer                    3.82        0.0        0.0   \n",
       "metastatic_pancreatic_adenocarcinoma            3.75        0.0        0.0   \n",
       "\n",
       "                                      sig_IL-1R  sig_IL-1RA  n_sig  max_score  \n",
       "token                                                                          \n",
       "relapsed_renal_cell_carcinoma               0.0        0.00      1       4.58  \n",
       "adca3                                       0.0        3.99      1       3.99  \n",
       "rp60                                        3.9        0.00      1       3.90  \n",
       "recurrent_endometrial_cancer                0.0        0.00      1       3.82  \n",
       "metastatic_pancreatic_adenocarcinoma        0.0        0.00      1       3.75  "
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_finalglobal = agg_cols(df_max_global_merged, 1.3)\n",
    "df_finalglobal.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_finalglobal.to_csv(\"./Final_allglobal_scores_combined.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "def stitch_pretokens(dataframe):\n",
    "    dataframe['pref_name_new'] = ''\n",
    "    PRESTAGING_COOKIE = {'csrftoken':'wLgolDsSbcb3Py2xB8xuiRDmwezQ9Y9o', 'sessionid':'jqab5fdwqmrxrnc5u0cli8xqxerep2p2'}  \n",
    "    mappings_dict = {}\n",
    "    tokens = dataframe.index.to_list()\n",
    "    for token in tokens:\n",
    "        print(token)\n",
    "        api = 'https://preview.nferx.com/dv/201912/nucleus-api/get_entities_for_token?'\n",
    "        params = {\n",
    "            'token' : token,\n",
    "            'is_good_synonym' : 1, \n",
    "            'entity_type' : 'disease'\n",
    "        }\n",
    "        try:\n",
    "            res = requests.get(api +  urlencode(params), cookies=PRESTAGING_COOKIE).json()['response']['records']\n",
    "            if len(res) > 1:\n",
    "                pref_tokens = [elem['pref_token'] for elem in res]\n",
    "                print(\"more than one pref_name returned %s: {}\"%(token, pref_tokens))\n",
    "            elif len(res) == 1:\n",
    "                pref_tokens = res[0]['pref_token']\n",
    "            else:\n",
    "                print('Nothing returned for token %s'%token)\n",
    "                pref_tokens = ''\n",
    "            dataframe.at[token, 'pref_name_new'] = pref_tokens\n",
    "        except Exception as e:\n",
    "            print(\"error from token: %s; %s\"%(token,e))\n",
    "#         if 'pref_token' in json_normalize(res.json()['response']['records']).columns:\n",
    "#             mappings_dict[dataframe.index[i]] = list(json_normalize(res.json()['response']['records'])['pref_token'])\n",
    "#         else:\n",
    "#             mappings_dict[dataframe.index[i]] = ['None']\n",
    "    return dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "nferx_py v1.7.4 is currently installed. Please upgrade to the latest version v2.0.0. Installation link: https://github.com/lumenbiomics/nferx_py\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2020-07-22 11:18:54,361 : INFO : URL: https://staging.nferx.com/api/is_authenticated, process-time: 1.5631420612335205\n",
      "2020-07-22 11:18:54,364 : INFO : Authentication successful - nferX is online\n",
      "2020-07-22 11:18:55,588 : INFO : URL: https://staging.nferx.com/api/get_data_versions, process-time: 1.2225661277770996\n",
      "2020-07-22 11:18:55,589 : INFO : Data version \"201912\" has been set as default one\n",
      "2020-07-22 11:18:55,591 : INFO : Default parameter server modified\n",
      "2020-07-22 11:18:55,592 : INFO : Default parameter api_server modified\n"
     ]
    }
   ],
   "source": [
    "import nferx_py.fn as nf\n",
    "from pymongo import MongoClient\n",
    "\n",
    "nf.authenticate('nfer','6eaa1d27bfa0639f2712191fa55df872')\n",
    "nf.modify_defaults('server', 'pre-staging')\n",
    "nf.modify_defaults('api_server', 'pre-staging')\n",
    "AUTH = nf.AUTH\n",
    "\n",
    "client = MongoClient('mongodb://deeksha:pg599ohismw1bio@mongo.servers.nferx.com:27017/')\n",
    "mongo_db_name='synonyms_db'\n",
    "mongo_db_collection='disease_synonyms'\n",
    "db = client[mongo_db_name]\n",
    "#coll = db[mongo_db_collection]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "table=db.disease_synonyms \n",
    "table.count_documents({}) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "first_instance = table.find_one()\n",
    "first_instance.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 169,
   "metadata": {},
   "outputs": [],
   "source": [
    "cursor = table.aggregate([\n",
    "{'$match': {'$and': [{'synonyms.token': {'$in': df_finalglobal.index.to_list()}},\n",
    "                     {'synonyms.attributes.good_synonym': True}]}},\n",
    "{'$project': {\n",
    "    'token': {'$filter': {'input': '$synonyms.token', 'as': 'syn',\n",
    "                             'cond': {'$in': ['$$syn', df_finalglobal.index.to_list()]}}},\n",
    "    'pref_name_token': 1,\n",
    "    '_id': 1\n",
    "\n",
    "}}])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 170,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "55980"
      ]
     },
     "execution_count": 170,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "table=db.disease_synonyms \n",
    "table.count_documents({}) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 171,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>_id</th>\n",
       "      <th>pref_name_token</th>\n",
       "      <th>token</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>7a98c3e1-be73-462a-b8f0-cc852cc2ea91</td>\n",
       "      <td>2_4_dienoyl_coa_reductase_deficiency</td>\n",
       "      <td>[decrd]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>dd4af5b3-a09e-41a2-8ffd-95fd2d8a8379</td>\n",
       "      <td>hmg_coa_lyase_deficiency</td>\n",
       "      <td>[3_hydroxy_3_methylglutaryl_coa_lyase_deficien...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>adcbd6aa-8125-4c44-8580-777415e719dd</td>\n",
       "      <td>46_xy_gonadal_dysgenesis_complete_sry_related</td>\n",
       "      <td>[srxy1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>2b789a00-d957-4cbe-bd9e-5fb9be8a5c7a</td>\n",
       "      <td>46_xy_sex_reversal_10</td>\n",
       "      <td>[srxy10]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>daa89efc-2272-4703-875c-fa6d1e380f56</td>\n",
       "      <td>46_xy_disorder_of_sex_development</td>\n",
       "      <td>[46_xy_dsd]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    _id  \\\n",
       "0  7a98c3e1-be73-462a-b8f0-cc852cc2ea91   \n",
       "1  dd4af5b3-a09e-41a2-8ffd-95fd2d8a8379   \n",
       "2  adcbd6aa-8125-4c44-8580-777415e719dd   \n",
       "3  2b789a00-d957-4cbe-bd9e-5fb9be8a5c7a   \n",
       "4  daa89efc-2272-4703-875c-fa6d1e380f56   \n",
       "\n",
       "                                 pref_name_token  \\\n",
       "0           2_4_dienoyl_coa_reductase_deficiency   \n",
       "1                       hmg_coa_lyase_deficiency   \n",
       "2  46_xy_gonadal_dysgenesis_complete_sry_related   \n",
       "3                          46_xy_sex_reversal_10   \n",
       "4              46_xy_disorder_of_sex_development   \n",
       "\n",
       "                                               token  \n",
       "0                                            [decrd]  \n",
       "1  [3_hydroxy_3_methylglutaryl_coa_lyase_deficien...  \n",
       "2                                            [srxy1]  \n",
       "3                                           [srxy10]  \n",
       "4                                        [46_xy_dsd]  "
      ]
     },
     "execution_count": 171,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pref_df = pd.DataFrame(cursor)\n",
    "pref_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys(['_id', 'pref_name', 'pref_name_token', 'synonyms', 'entity', 'relations'])"
      ]
     },
     "execution_count": 96,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "first_instance = table.find_one()\n",
    "first_instance.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'2_4_dienoyl_coa_reductase_deficiency'"
      ]
     },
     "execution_count": 100,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "first_instance['pref_name']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 157,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'name': '4_dienoyl_coa_reductase_deficiency_2',\n",
       "  'token': '4_dienoyl_coa_reductase_deficiency_2',\n",
       "  'type': ['synonym'],\n",
       "  'is_synonym': True,\n",
       "  'sources': {'UNIPROT_DISEASE': {'93e19430-7449-4460-a177-c702ec6da40a': ['PREF_NAME']},\n",
       "   'GARD': {'1e0e4587-f638-455a-b3aa-d99366ee3ffb': ['PREF_NAME']},\n",
       "   'ORPHANET': {'ac109ed0-ac25-4935-a3c6-458359bcc828': ['SYNONYM']}},\n",
       "  'original_tokens': [],\n",
       "  'normalized_variants': ['4_dienoyl_coa_reductase_deficiency_2'],\n",
       "  'attributes': {'occurrence_count': 0,\n",
       "   'pubmed_document_count': 0,\n",
       "   'document_count': 0,\n",
       "   'ctrials_document_count': 0,\n",
       "   'uspto_document_count': 0,\n",
       "   'pubmed_concurrence_docs': 0,\n",
       "   'concurrence_document_count': 0,\n",
       "   'p(e|t)_pubmed': 0,\n",
       "   'p(t|e)_pubmed': 0,\n",
       "   'ctrials_concurrence_docs': 0,\n",
       "   'p(e|t)_ctrials': 0,\n",
       "   'p(t|e)_ctrials': 0,\n",
       "   'uspto_concurrence_docs': 0,\n",
       "   'p(e|t)_uspto': 0,\n",
       "   'p(t|e)_uspto': 0,\n",
       "   'source_count': 3,\n",
       "   'concurrence_ontology_count': 3,\n",
       "   'p(t|e)_ontologies': 0.75,\n",
       "   'calculated_syn_weight': 0,\n",
       "   'good_synonym': True,\n",
       "   'logic': 'DUAL_PHRASE_VARIANT',\n",
       "   'p(t|e)_corpus': 0,\n",
       "   'p(e|t)_corpus': 0,\n",
       "   'relevance_to_entity': 0}},\n",
       " {'name': '2_4_dienoyl_coa_reductase_deficiency',\n",
       "  'token': '2_4_dienoyl_coa_reductase_deficiency',\n",
       "  'type': ['synonym', 'pref_name'],\n",
       "  'is_synonym': True,\n",
       "  'sources': {'UNIPROT_DISEASE': {'93e19430-7449-4460-a177-c702ec6da40a': ['PREF_NAME']},\n",
       "   'GARD': {'1e0e4587-f638-455a-b3aa-d99366ee3ffb': ['PREF_NAME']},\n",
       "   'ORPHANET': {'ac109ed0-ac25-4935-a3c6-458359bcc828': ['SYNONYM']},\n",
       "   'UMLS': {'69cd80b8-612f-4b6b-8719-5d7c11e68197': ['PREF_NAME']}},\n",
       "  'original_tokens': [],\n",
       "  'normalized_variants': ['2_4_dienoyl_coa_reductase_deficiency'],\n",
       "  'attributes': {'occurrence_count': 66,\n",
       "   'pubmed_document_count': 2,\n",
       "   'document_count': 2,\n",
       "   'ctrials_document_count': 0,\n",
       "   'uspto_document_count': 0,\n",
       "   'pubmed_concurrence_docs': 2,\n",
       "   'concurrence_document_count': 2,\n",
       "   'p(e|t)_pubmed': 1,\n",
       "   'p(t|e)_pubmed': 1,\n",
       "   'ctrials_concurrence_docs': 0,\n",
       "   'p(e|t)_ctrials': 0,\n",
       "   'p(t|e)_ctrials': 0,\n",
       "   'uspto_concurrence_docs': 0,\n",
       "   'p(e|t)_uspto': 0,\n",
       "   'p(t|e)_uspto': 0,\n",
       "   'source_count': 4,\n",
       "   'concurrence_ontology_count': 4,\n",
       "   'p(t|e)_ontologies': 1,\n",
       "   'calculated_syn_weight': 1,\n",
       "   'good_synonym': True,\n",
       "   'logic': 'PREF_TOKEN',\n",
       "   'p(t|e)_corpus': 0,\n",
       "   'p(e|t)_corpus': 0,\n",
       "   'relevance_to_entity': 1}},\n",
       " {'name': 'dienoyl_coa_reductase_deficiency',\n",
       "  'token': 'dienoyl_coa_reductase_deficiency',\n",
       "  'type': ['synonym'],\n",
       "  'is_synonym': True,\n",
       "  'sources': {'GARD': {'1e0e4587-f638-455a-b3aa-d99366ee3ffb': ['SYNONYM']}},\n",
       "  'original_tokens': [],\n",
       "  'normalized_variants': ['dienoyl_coa_reductase_deficiency'],\n",
       "  'attributes': {'occurrence_count': 647,\n",
       "   'pubmed_document_count': 10,\n",
       "   'document_count': 36,\n",
       "   'ctrials_document_count': 0,\n",
       "   'uspto_document_count': 26,\n",
       "   'pubmed_concurrence_docs': 0,\n",
       "   'concurrence_document_count': 0,\n",
       "   'p(e|t)_pubmed': 0,\n",
       "   'p(t|e)_pubmed': 0,\n",
       "   'ctrials_concurrence_docs': 0,\n",
       "   'p(e|t)_ctrials': 0,\n",
       "   'p(t|e)_ctrials': 0,\n",
       "   'uspto_concurrence_docs': 0,\n",
       "   'p(e|t)_uspto': 0,\n",
       "   'p(t|e)_uspto': 0,\n",
       "   'source_count': 1,\n",
       "   'concurrence_ontology_count': 1,\n",
       "   'p(t|e)_ontologies': 0.25,\n",
       "   'calculated_syn_weight': 0,\n",
       "   'good_synonym': False,\n",
       "   'logic': 'NEW_ENTITY_SCORE',\n",
       "   'p(t|e)_corpus': 0,\n",
       "   'p(e|t)_corpus': 0,\n",
       "   'relevance_to_entity': 0}},\n",
       " {'name': '4_alpha_dienoyl_coa_reductase_deficiency_2',\n",
       "  'token': '4_alpha_dienoyl_coa_reductase_deficiency_2',\n",
       "  'type': ['synonym'],\n",
       "  'is_synonym': True,\n",
       "  'sources': {'GARD': {'1e0e4587-f638-455a-b3aa-d99366ee3ffb': ['SYNONYM']}},\n",
       "  'original_tokens': [],\n",
       "  'normalized_variants': ['4_alpha_dienoyl_coa_reductase_deficiency_2'],\n",
       "  'attributes': {'occurrence_count': 0,\n",
       "   'pubmed_document_count': 0,\n",
       "   'document_count': 0,\n",
       "   'ctrials_document_count': 0,\n",
       "   'uspto_document_count': 0,\n",
       "   'pubmed_concurrence_docs': 0,\n",
       "   'concurrence_document_count': 0,\n",
       "   'p(e|t)_pubmed': 0,\n",
       "   'p(t|e)_pubmed': 0,\n",
       "   'ctrials_concurrence_docs': 0,\n",
       "   'p(e|t)_ctrials': 0,\n",
       "   'p(t|e)_ctrials': 0,\n",
       "   'uspto_concurrence_docs': 0,\n",
       "   'p(e|t)_uspto': 0,\n",
       "   'p(t|e)_uspto': 0,\n",
       "   'source_count': 1,\n",
       "   'concurrence_ontology_count': 1,\n",
       "   'p(t|e)_ontologies': 0.25,\n",
       "   'calculated_syn_weight': 0,\n",
       "   'good_synonym': False,\n",
       "   'logic': 'NEW_ENTITY_SCORE',\n",
       "   'p(t|e)_corpus': 0,\n",
       "   'p(e|t)_corpus': 0,\n",
       "   'relevance_to_entity': 0}},\n",
       " {'name': '2_4_alpha_dienoyl_coa_reductase_deficiency',\n",
       "  'token': '2_4_alpha_dienoyl_coa_reductase_deficiency',\n",
       "  'type': ['synonym'],\n",
       "  'is_synonym': True,\n",
       "  'sources': {'GARD': {'1e0e4587-f638-455a-b3aa-d99366ee3ffb': ['SYNONYM']}},\n",
       "  'original_tokens': [],\n",
       "  'normalized_variants': ['2_4_alpha_dienoyl_coa_reductase_deficiency'],\n",
       "  'attributes': {'occurrence_count': 0,\n",
       "   'pubmed_document_count': 0,\n",
       "   'document_count': 0,\n",
       "   'ctrials_document_count': 0,\n",
       "   'uspto_document_count': 0,\n",
       "   'pubmed_concurrence_docs': 0,\n",
       "   'concurrence_document_count': 0,\n",
       "   'p(e|t)_pubmed': 0,\n",
       "   'p(t|e)_pubmed': 0,\n",
       "   'ctrials_concurrence_docs': 0,\n",
       "   'p(e|t)_ctrials': 0,\n",
       "   'p(t|e)_ctrials': 0,\n",
       "   'uspto_concurrence_docs': 0,\n",
       "   'p(e|t)_uspto': 0,\n",
       "   'p(t|e)_uspto': 0,\n",
       "   'source_count': 1,\n",
       "   'concurrence_ontology_count': 1,\n",
       "   'p(t|e)_ontologies': 0.25,\n",
       "   'calculated_syn_weight': 0,\n",
       "   'good_synonym': False,\n",
       "   'logic': 'NEW_ENTITY_SCORE',\n",
       "   'p(t|e)_corpus': 0,\n",
       "   'p(e|t)_corpus': 0,\n",
       "   'relevance_to_entity': 0}},\n",
       " {'name': 'progressive_encephalopathy_with_leukodystrophy_due_to_decr_deficiency',\n",
       "  'token': 'progressive_encephalopathy_with_leukodystrophy_due_to_decr_deficiency',\n",
       "  'type': ['synonym'],\n",
       "  'is_synonym': True,\n",
       "  'sources': {'ORPHANET': {'ac109ed0-ac25-4935-a3c6-458359bcc828': ['PREF_NAME']}},\n",
       "  'original_tokens': [],\n",
       "  'normalized_variants': ['progressive_encephalopathy_with_leukodystrophy_due_to_decr_deficiency'],\n",
       "  'attributes': {'occurrence_count': 18,\n",
       "   'pubmed_document_count': 0,\n",
       "   'document_count': 0,\n",
       "   'ctrials_document_count': 0,\n",
       "   'uspto_document_count': 0,\n",
       "   'pubmed_concurrence_docs': 0,\n",
       "   'concurrence_document_count': 0,\n",
       "   'p(e|t)_pubmed': 0,\n",
       "   'p(t|e)_pubmed': 0,\n",
       "   'ctrials_concurrence_docs': 0,\n",
       "   'p(e|t)_ctrials': 0,\n",
       "   'p(t|e)_ctrials': 0,\n",
       "   'uspto_concurrence_docs': 0,\n",
       "   'p(e|t)_uspto': 0,\n",
       "   'p(t|e)_uspto': 0,\n",
       "   'source_count': 1,\n",
       "   'concurrence_ontology_count': 1,\n",
       "   'p(t|e)_ontologies': 0.25,\n",
       "   'calculated_syn_weight': 0,\n",
       "   'good_synonym': False,\n",
       "   'logic': 'NEW_ENTITY_SCORE',\n",
       "   'p(t|e)_corpus': 0,\n",
       "   'p(e|t)_corpus': 0,\n",
       "   'relevance_to_entity': 0}},\n",
       " {'name': 'decr_deficiency_with_hyperlysinemia',\n",
       "  'token': 'decr_deficiency_with_hyperlysinemia',\n",
       "  'type': ['synonym'],\n",
       "  'is_synonym': True,\n",
       "  'sources': {'ORPHANET': {'ac109ed0-ac25-4935-a3c6-458359bcc828': ['SYNONYM']}},\n",
       "  'original_tokens': [],\n",
       "  'normalized_variants': ['decr_deficiency_with_hyperlysinemia'],\n",
       "  'attributes': {'occurrence_count': 15,\n",
       "   'pubmed_document_count': 1,\n",
       "   'document_count': 1,\n",
       "   'ctrials_document_count': 0,\n",
       "   'uspto_document_count': 0,\n",
       "   'pubmed_concurrence_docs': 0,\n",
       "   'concurrence_document_count': 0,\n",
       "   'p(e|t)_pubmed': 0,\n",
       "   'p(t|e)_pubmed': 0,\n",
       "   'ctrials_concurrence_docs': 0,\n",
       "   'p(e|t)_ctrials': 0,\n",
       "   'p(t|e)_ctrials': 0,\n",
       "   'uspto_concurrence_docs': 0,\n",
       "   'p(e|t)_uspto': 0,\n",
       "   'p(t|e)_uspto': 0,\n",
       "   'source_count': 1,\n",
       "   'concurrence_ontology_count': 1,\n",
       "   'p(t|e)_ontologies': 0.25,\n",
       "   'calculated_syn_weight': 0,\n",
       "   'good_synonym': False,\n",
       "   'logic': 'NEW_ENTITY_SCORE',\n",
       "   'p(t|e)_corpus': 0,\n",
       "   'p(e|t)_corpus': 0,\n",
       "   'relevance_to_entity': 0}},\n",
       " {'name': 'decrd',\n",
       "  'token': 'decrd',\n",
       "  'type': ['synonym'],\n",
       "  'is_synonym': True,\n",
       "  'sources': {'UMLS': {'69cd80b8-612f-4b6b-8719-5d7c11e68197': ['SYNONYM']}},\n",
       "  'original_tokens': [],\n",
       "  'normalized_variants': ['decrd'],\n",
       "  'attributes': {'occurrence_count': 379,\n",
       "   'pubmed_document_count': 2,\n",
       "   'document_count': 3,\n",
       "   'ctrials_document_count': 0,\n",
       "   'uspto_document_count': 1,\n",
       "   'pubmed_concurrence_docs': 0,\n",
       "   'concurrence_document_count': 0,\n",
       "   'p(e|t)_pubmed': 0,\n",
       "   'p(t|e)_pubmed': 0,\n",
       "   'ctrials_concurrence_docs': 0,\n",
       "   'p(e|t)_ctrials': 0,\n",
       "   'p(t|e)_ctrials': 0,\n",
       "   'uspto_concurrence_docs': 0,\n",
       "   'p(e|t)_uspto': 0,\n",
       "   'p(t|e)_uspto': 0,\n",
       "   'source_count': 1,\n",
       "   'concurrence_ontology_count': 1,\n",
       "   'p(t|e)_ontologies': 0.25,\n",
       "   'relevance_across_entities': 0.5,\n",
       "   'calculated_syn_weight': 0,\n",
       "   'good_synonym': False,\n",
       "   'logic': 'NEW_ENTITY_SCORE',\n",
       "   'p(t|e)_corpus': 0,\n",
       "   'p(e|t)_corpus': 0,\n",
       "   'relevance_to_entity': 0}}]"
      ]
     },
     "execution_count": 157,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "first_instance['synonyms']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'name': 'progressive_encephalopathy_with_leukodystrophy_due_to_decr_deficiency',\n",
       "  'token': 'progressive_encephalopathy_with_leukodystrophy_due_to_decr_deficiency',\n",
       "  'type': ['synonym'],\n",
       "  'is_synonym': True,\n",
       "  'sources': {'ORPHANET': {'ac109ed0-ac25-4935-a3c6-458359bcc828': ['PREF_NAME']}},\n",
       "  'original_tokens': [],\n",
       "  'normalized_variants': ['progressive_encephalopathy_with_leukodystrophy_due_to_decr_deficiency'],\n",
       "  'attributes': {'occurrence_count': 18,\n",
       "   'pubmed_document_count': 0,\n",
       "   'document_count': 0,\n",
       "   'ctrials_document_count': 0,\n",
       "   'uspto_document_count': 0,\n",
       "   'pubmed_concurrence_docs': 0,\n",
       "   'concurrence_document_count': 0,\n",
       "   'p(e|t)_pubmed': 0,\n",
       "   'p(t|e)_pubmed': 0,\n",
       "   'ctrials_concurrence_docs': 0,\n",
       "   'p(e|t)_ctrials': 0,\n",
       "   'p(t|e)_ctrials': 0,\n",
       "   'uspto_concurrence_docs': 0,\n",
       "   'p(e|t)_uspto': 0,\n",
       "   'p(t|e)_uspto': 0,\n",
       "   'source_count': 1,\n",
       "   'concurrence_ontology_count': 1,\n",
       "   'p(t|e)_ontologies': 0.25,\n",
       "   'calculated_syn_weight': 0,\n",
       "   'good_synonym': False,\n",
       "   'logic': 'NEW_ENTITY_SCORE',\n",
       "   'p(t|e)_corpus': 0,\n",
       "   'p(e|t)_corpus': 0,\n",
       "   'relevance_to_entity': 0}}]"
      ]
     },
     "execution_count": 135,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(filter(lambda match: match['name'] == 'progressive_encephalopathy_with_leukodystrophy_due_to_decr_deficiency', table.find_one({'pref_name':'2_4_dienoyl_coa_reductase_deficiency'})['synonyms']))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 166,
   "metadata": {},
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "invalid syntax (<ipython-input-166-03d28934d5e5>, line 2)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;36m  File \u001b[0;32m\"<ipython-input-166-03d28934d5e5>\"\u001b[0;36m, line \u001b[0;32m2\u001b[0m\n\u001b[0;31m    synonyms: { $elemMatch: { token: \"dienoyl_coa_reductase_deficiency\"}}\u001b[0m\n\u001b[0m                ^\u001b[0m\n\u001b[0;31mSyntaxError\u001b[0m\u001b[0;31m:\u001b[0m invalid syntax\n"
     ]
    }
   ],
   "source": [
    "table.find({\n",
    "  synonyms: {$elemMatch: { token: \"dienoyl_coa_reductase_deficiency\"}}\n",
    "})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'name': '4_dienoyl_coa_reductase_deficiency_2',\n",
       "  'token': '4_dienoyl_coa_reductase_deficiency_2',\n",
       "  'type': ['synonym'],\n",
       "  'is_synonym': True,\n",
       "  'sources': {'UNIPROT_DISEASE': {'93e19430-7449-4460-a177-c702ec6da40a': ['PREF_NAME']},\n",
       "   'GARD': {'1e0e4587-f638-455a-b3aa-d99366ee3ffb': ['PREF_NAME']},\n",
       "   'ORPHANET': {'ac109ed0-ac25-4935-a3c6-458359bcc828': ['SYNONYM']}},\n",
       "  'original_tokens': [],\n",
       "  'normalized_variants': ['4_dienoyl_coa_reductase_deficiency_2'],\n",
       "  'attributes': {'occurrence_count': 0,\n",
       "   'pubmed_document_count': 0,\n",
       "   'document_count': 0,\n",
       "   'ctrials_document_count': 0,\n",
       "   'uspto_document_count': 0,\n",
       "   'pubmed_concurrence_docs': 0,\n",
       "   'concurrence_document_count': 0,\n",
       "   'p(e|t)_pubmed': 0,\n",
       "   'p(t|e)_pubmed': 0,\n",
       "   'ctrials_concurrence_docs': 0,\n",
       "   'p(e|t)_ctrials': 0,\n",
       "   'p(t|e)_ctrials': 0,\n",
       "   'uspto_concurrence_docs': 0,\n",
       "   'p(e|t)_uspto': 0,\n",
       "   'p(t|e)_uspto': 0,\n",
       "   'source_count': 3,\n",
       "   'concurrence_ontology_count': 3,\n",
       "   'p(t|e)_ontologies': 0.75,\n",
       "   'calculated_syn_weight': 0,\n",
       "   'good_synonym': True,\n",
       "   'logic': 'DUAL_PHRASE_VARIANT',\n",
       "   'p(t|e)_corpus': 0,\n",
       "   'p(e|t)_corpus': 0,\n",
       "   'relevance_to_entity': 0}},\n",
       " {'name': '2_4_dienoyl_coa_reductase_deficiency',\n",
       "  'token': '2_4_dienoyl_coa_reductase_deficiency',\n",
       "  'type': ['synonym', 'pref_name'],\n",
       "  'is_synonym': True,\n",
       "  'sources': {'UNIPROT_DISEASE': {'93e19430-7449-4460-a177-c702ec6da40a': ['PREF_NAME']},\n",
       "   'GARD': {'1e0e4587-f638-455a-b3aa-d99366ee3ffb': ['PREF_NAME']},\n",
       "   'ORPHANET': {'ac109ed0-ac25-4935-a3c6-458359bcc828': ['SYNONYM']},\n",
       "   'UMLS': {'69cd80b8-612f-4b6b-8719-5d7c11e68197': ['PREF_NAME']}},\n",
       "  'original_tokens': [],\n",
       "  'normalized_variants': ['2_4_dienoyl_coa_reductase_deficiency'],\n",
       "  'attributes': {'occurrence_count': 66,\n",
       "   'pubmed_document_count': 2,\n",
       "   'document_count': 2,\n",
       "   'ctrials_document_count': 0,\n",
       "   'uspto_document_count': 0,\n",
       "   'pubmed_concurrence_docs': 2,\n",
       "   'concurrence_document_count': 2,\n",
       "   'p(e|t)_pubmed': 1,\n",
       "   'p(t|e)_pubmed': 1,\n",
       "   'ctrials_concurrence_docs': 0,\n",
       "   'p(e|t)_ctrials': 0,\n",
       "   'p(t|e)_ctrials': 0,\n",
       "   'uspto_concurrence_docs': 0,\n",
       "   'p(e|t)_uspto': 0,\n",
       "   'p(t|e)_uspto': 0,\n",
       "   'source_count': 4,\n",
       "   'concurrence_ontology_count': 4,\n",
       "   'p(t|e)_ontologies': 1,\n",
       "   'calculated_syn_weight': 1,\n",
       "   'good_synonym': True,\n",
       "   'logic': 'PREF_TOKEN',\n",
       "   'p(t|e)_corpus': 0,\n",
       "   'p(e|t)_corpus': 0,\n",
       "   'relevance_to_entity': 1}},\n",
       " {'name': 'dienoyl_coa_reductase_deficiency',\n",
       "  'token': 'dienoyl_coa_reductase_deficiency',\n",
       "  'type': ['synonym'],\n",
       "  'is_synonym': True,\n",
       "  'sources': {'GARD': {'1e0e4587-f638-455a-b3aa-d99366ee3ffb': ['SYNONYM']}},\n",
       "  'original_tokens': [],\n",
       "  'normalized_variants': ['dienoyl_coa_reductase_deficiency'],\n",
       "  'attributes': {'occurrence_count': 647,\n",
       "   'pubmed_document_count': 10,\n",
       "   'document_count': 36,\n",
       "   'ctrials_document_count': 0,\n",
       "   'uspto_document_count': 26,\n",
       "   'pubmed_concurrence_docs': 0,\n",
       "   'concurrence_document_count': 0,\n",
       "   'p(e|t)_pubmed': 0,\n",
       "   'p(t|e)_pubmed': 0,\n",
       "   'ctrials_concurrence_docs': 0,\n",
       "   'p(e|t)_ctrials': 0,\n",
       "   'p(t|e)_ctrials': 0,\n",
       "   'uspto_concurrence_docs': 0,\n",
       "   'p(e|t)_uspto': 0,\n",
       "   'p(t|e)_uspto': 0,\n",
       "   'source_count': 1,\n",
       "   'concurrence_ontology_count': 1,\n",
       "   'p(t|e)_ontologies': 0.25,\n",
       "   'calculated_syn_weight': 0,\n",
       "   'good_synonym': False,\n",
       "   'logic': 'NEW_ENTITY_SCORE',\n",
       "   'p(t|e)_corpus': 0,\n",
       "   'p(e|t)_corpus': 0,\n",
       "   'relevance_to_entity': 0}},\n",
       " {'name': '4_alpha_dienoyl_coa_reductase_deficiency_2',\n",
       "  'token': '4_alpha_dienoyl_coa_reductase_deficiency_2',\n",
       "  'type': ['synonym'],\n",
       "  'is_synonym': True,\n",
       "  'sources': {'GARD': {'1e0e4587-f638-455a-b3aa-d99366ee3ffb': ['SYNONYM']}},\n",
       "  'original_tokens': [],\n",
       "  'normalized_variants': ['4_alpha_dienoyl_coa_reductase_deficiency_2'],\n",
       "  'attributes': {'occurrence_count': 0,\n",
       "   'pubmed_document_count': 0,\n",
       "   'document_count': 0,\n",
       "   'ctrials_document_count': 0,\n",
       "   'uspto_document_count': 0,\n",
       "   'pubmed_concurrence_docs': 0,\n",
       "   'concurrence_document_count': 0,\n",
       "   'p(e|t)_pubmed': 0,\n",
       "   'p(t|e)_pubmed': 0,\n",
       "   'ctrials_concurrence_docs': 0,\n",
       "   'p(e|t)_ctrials': 0,\n",
       "   'p(t|e)_ctrials': 0,\n",
       "   'uspto_concurrence_docs': 0,\n",
       "   'p(e|t)_uspto': 0,\n",
       "   'p(t|e)_uspto': 0,\n",
       "   'source_count': 1,\n",
       "   'concurrence_ontology_count': 1,\n",
       "   'p(t|e)_ontologies': 0.25,\n",
       "   'calculated_syn_weight': 0,\n",
       "   'good_synonym': False,\n",
       "   'logic': 'NEW_ENTITY_SCORE',\n",
       "   'p(t|e)_corpus': 0,\n",
       "   'p(e|t)_corpus': 0,\n",
       "   'relevance_to_entity': 0}},\n",
       " {'name': '2_4_alpha_dienoyl_coa_reductase_deficiency',\n",
       "  'token': '2_4_alpha_dienoyl_coa_reductase_deficiency',\n",
       "  'type': ['synonym'],\n",
       "  'is_synonym': True,\n",
       "  'sources': {'GARD': {'1e0e4587-f638-455a-b3aa-d99366ee3ffb': ['SYNONYM']}},\n",
       "  'original_tokens': [],\n",
       "  'normalized_variants': ['2_4_alpha_dienoyl_coa_reductase_deficiency'],\n",
       "  'attributes': {'occurrence_count': 0,\n",
       "   'pubmed_document_count': 0,\n",
       "   'document_count': 0,\n",
       "   'ctrials_document_count': 0,\n",
       "   'uspto_document_count': 0,\n",
       "   'pubmed_concurrence_docs': 0,\n",
       "   'concurrence_document_count': 0,\n",
       "   'p(e|t)_pubmed': 0,\n",
       "   'p(t|e)_pubmed': 0,\n",
       "   'ctrials_concurrence_docs': 0,\n",
       "   'p(e|t)_ctrials': 0,\n",
       "   'p(t|e)_ctrials': 0,\n",
       "   'uspto_concurrence_docs': 0,\n",
       "   'p(e|t)_uspto': 0,\n",
       "   'p(t|e)_uspto': 0,\n",
       "   'source_count': 1,\n",
       "   'concurrence_ontology_count': 1,\n",
       "   'p(t|e)_ontologies': 0.25,\n",
       "   'calculated_syn_weight': 0,\n",
       "   'good_synonym': False,\n",
       "   'logic': 'NEW_ENTITY_SCORE',\n",
       "   'p(t|e)_corpus': 0,\n",
       "   'p(e|t)_corpus': 0,\n",
       "   'relevance_to_entity': 0}},\n",
       " {'name': 'progressive_encephalopathy_with_leukodystrophy_due_to_decr_deficiency',\n",
       "  'token': 'progressive_encephalopathy_with_leukodystrophy_due_to_decr_deficiency',\n",
       "  'type': ['synonym'],\n",
       "  'is_synonym': True,\n",
       "  'sources': {'ORPHANET': {'ac109ed0-ac25-4935-a3c6-458359bcc828': ['PREF_NAME']}},\n",
       "  'original_tokens': [],\n",
       "  'normalized_variants': ['progressive_encephalopathy_with_leukodystrophy_due_to_decr_deficiency'],\n",
       "  'attributes': {'occurrence_count': 18,\n",
       "   'pubmed_document_count': 0,\n",
       "   'document_count': 0,\n",
       "   'ctrials_document_count': 0,\n",
       "   'uspto_document_count': 0,\n",
       "   'pubmed_concurrence_docs': 0,\n",
       "   'concurrence_document_count': 0,\n",
       "   'p(e|t)_pubmed': 0,\n",
       "   'p(t|e)_pubmed': 0,\n",
       "   'ctrials_concurrence_docs': 0,\n",
       "   'p(e|t)_ctrials': 0,\n",
       "   'p(t|e)_ctrials': 0,\n",
       "   'uspto_concurrence_docs': 0,\n",
       "   'p(e|t)_uspto': 0,\n",
       "   'p(t|e)_uspto': 0,\n",
       "   'source_count': 1,\n",
       "   'concurrence_ontology_count': 1,\n",
       "   'p(t|e)_ontologies': 0.25,\n",
       "   'calculated_syn_weight': 0,\n",
       "   'good_synonym': False,\n",
       "   'logic': 'NEW_ENTITY_SCORE',\n",
       "   'p(t|e)_corpus': 0,\n",
       "   'p(e|t)_corpus': 0,\n",
       "   'relevance_to_entity': 0}},\n",
       " {'name': 'decr_deficiency_with_hyperlysinemia',\n",
       "  'token': 'decr_deficiency_with_hyperlysinemia',\n",
       "  'type': ['synonym'],\n",
       "  'is_synonym': True,\n",
       "  'sources': {'ORPHANET': {'ac109ed0-ac25-4935-a3c6-458359bcc828': ['SYNONYM']}},\n",
       "  'original_tokens': [],\n",
       "  'normalized_variants': ['decr_deficiency_with_hyperlysinemia'],\n",
       "  'attributes': {'occurrence_count': 15,\n",
       "   'pubmed_document_count': 1,\n",
       "   'document_count': 1,\n",
       "   'ctrials_document_count': 0,\n",
       "   'uspto_document_count': 0,\n",
       "   'pubmed_concurrence_docs': 0,\n",
       "   'concurrence_document_count': 0,\n",
       "   'p(e|t)_pubmed': 0,\n",
       "   'p(t|e)_pubmed': 0,\n",
       "   'ctrials_concurrence_docs': 0,\n",
       "   'p(e|t)_ctrials': 0,\n",
       "   'p(t|e)_ctrials': 0,\n",
       "   'uspto_concurrence_docs': 0,\n",
       "   'p(e|t)_uspto': 0,\n",
       "   'p(t|e)_uspto': 0,\n",
       "   'source_count': 1,\n",
       "   'concurrence_ontology_count': 1,\n",
       "   'p(t|e)_ontologies': 0.25,\n",
       "   'calculated_syn_weight': 0,\n",
       "   'good_synonym': False,\n",
       "   'logic': 'NEW_ENTITY_SCORE',\n",
       "   'p(t|e)_corpus': 0,\n",
       "   'p(e|t)_corpus': 0,\n",
       "   'relevance_to_entity': 0}},\n",
       " {'name': 'decrd',\n",
       "  'token': 'decrd',\n",
       "  'type': ['synonym'],\n",
       "  'is_synonym': True,\n",
       "  'sources': {'UMLS': {'69cd80b8-612f-4b6b-8719-5d7c11e68197': ['SYNONYM']}},\n",
       "  'original_tokens': [],\n",
       "  'normalized_variants': ['decrd'],\n",
       "  'attributes': {'occurrence_count': 379,\n",
       "   'pubmed_document_count': 2,\n",
       "   'document_count': 3,\n",
       "   'ctrials_document_count': 0,\n",
       "   'uspto_document_count': 1,\n",
       "   'pubmed_concurrence_docs': 0,\n",
       "   'concurrence_document_count': 0,\n",
       "   'p(e|t)_pubmed': 0,\n",
       "   'p(t|e)_pubmed': 0,\n",
       "   'ctrials_concurrence_docs': 0,\n",
       "   'p(e|t)_ctrials': 0,\n",
       "   'p(t|e)_ctrials': 0,\n",
       "   'uspto_concurrence_docs': 0,\n",
       "   'p(e|t)_uspto': 0,\n",
       "   'p(t|e)_uspto': 0,\n",
       "   'source_count': 1,\n",
       "   'concurrence_ontology_count': 1,\n",
       "   'p(t|e)_ontologies': 0.25,\n",
       "   'relevance_across_entities': 0.5,\n",
       "   'calculated_syn_weight': 0,\n",
       "   'good_synonym': False,\n",
       "   'logic': 'NEW_ENTITY_SCORE',\n",
       "   'p(t|e)_corpus': 0,\n",
       "   'p(e|t)_corpus': 0,\n",
       "   'relevance_to_entity': 0}}]"
      ]
     },
     "execution_count": 131,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "table.find_one({'pref_name':'2_4_dienoyl_coa_reductase_deficiency'})['synonyms']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'2_4_dienoyl_coa_reductase_deficiency'"
      ]
     },
     "execution_count": 109,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "first_instance['pref_name']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name '_ANY_' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-120-721920be1b71>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mtable\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mthings\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfind\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m{\u001b[0m\u001b[0m_ANY_\u001b[0m \u001b[0;34m:\u001b[0m \u001b[0;34m\"decr_deficiency_with_hyperlysinemia\"\u001b[0m\u001b[0;34m}\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m: name '_ANY_' is not defined"
     ]
    }
   ],
   "source": [
    "table.things.find({_ANY_ : \"decr_deficiency_with_hyperlysinemia\"}) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Collection(Database(MongoClient(host=['mongo.servers.nferx.com:27017'], document_class=dict, tz_aware=False, connect=True), 'synonyms_db'), 'disease_synonyms.pref_name')"
      ]
     },
     "execution_count": 121,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "table['pref_name']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "metadata": {},
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "list indices must be integers or slices, not str",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-128-51a9d935c353>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0mall_synonyms_dic\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m{\u001b[0m\u001b[0;34m}\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mdoc\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mtable\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfind\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 3\u001b[0;31m     \u001b[0mall_synonyms_dic\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mdoc\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'pref_name'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdoc\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'synonyms'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'name'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m: list indices must be integers or slices, not str"
     ]
    }
   ],
   "source": [
    "all_synonyms_dic = {}\n",
    "for doc in table.find():\n",
    "    all_synonyms_dic[doc['pref_name']] = doc['synonyms']['name']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "list(filter(lambda match: match['name'] == 'progressive_encephalopathy_with_leukodystrophy_due_to_decr_deficiency', table.find_one({'pref_name':'2_4_dienoyl_coa_reductase_deficiency'})['synonyms']))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-136-ef5256fb7e82>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;34m'21_hydroxylase_deficiency'\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 3\u001b[0;31m \u001b[0;32mfor\u001b[0m \u001b[0mdoc\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mtable\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfind\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      4\u001b[0m     if len(list(\n\u001b[1;32m      5\u001b[0m         filter(lambda match: match['name'] == '21_hydroxylase_deficiency', doc['synonyms']))) > 1:\n",
      "\u001b[0;32m~/anaconda3/lib/python3.7/site-packages/pymongo/cursor.py\u001b[0m in \u001b[0;36mnext\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m   1154\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__empty\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1155\u001b[0m             \u001b[0;32mraise\u001b[0m \u001b[0mStopIteration\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1156\u001b[0;31m         \u001b[0;32mif\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__data\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mor\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_refresh\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1157\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__manipulate\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1158\u001b[0m                 \u001b[0m_db\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__collection\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdatabase\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/lib/python3.7/site-packages/pymongo/cursor.py\u001b[0m in \u001b[0;36m_refresh\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m   1091\u001b[0m                                     \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__max_await_time_ms\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1092\u001b[0m                                     self.__exhaust_mgr)\n\u001b[0;32m-> 1093\u001b[0;31m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__send_message\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mg\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1094\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1095\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__data\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/lib/python3.7/site-packages/pymongo/cursor.py\u001b[0m in \u001b[0;36m__send_message\u001b[0;34m(self, operation)\u001b[0m\n\u001b[1;32m    953\u001b[0m             response = client._run_operation_with_response(\n\u001b[1;32m    954\u001b[0m                 \u001b[0moperation\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_unpack_response\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mexhaust\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__exhaust\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 955\u001b[0;31m                 address=self.__address)\n\u001b[0m\u001b[1;32m    956\u001b[0m         \u001b[0;32mexcept\u001b[0m \u001b[0mOperationFailure\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    957\u001b[0m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__killed\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mTrue\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/lib/python3.7/site-packages/pymongo/mongo_client.py\u001b[0m in \u001b[0;36m_run_operation_with_response\u001b[0;34m(self, operation, unpack_res, exhaust, address)\u001b[0m\n\u001b[1;32m   1344\u001b[0m             \u001b[0maddress\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0maddress\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1345\u001b[0m             \u001b[0mretryable\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0moperation\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmessage\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_Query\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1346\u001b[0;31m             exhaust=exhaust)\n\u001b[0m\u001b[1;32m   1347\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1348\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m_retry_with_session\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mretryable\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfunc\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msession\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mbulk\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/lib/python3.7/site-packages/pymongo/mongo_client.py\u001b[0m in \u001b[0;36m_retryable_read\u001b[0;34m(self, func, read_pref, session, address, retryable, exhaust)\u001b[0m\n\u001b[1;32m   1462\u001b[0m                         \u001b[0;31m# not support retryable reads, raise the last error.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1463\u001b[0m                         \u001b[0;32mraise\u001b[0m \u001b[0mlast_error\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1464\u001b[0;31m                     \u001b[0;32mreturn\u001b[0m \u001b[0mfunc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msession\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mserver\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msock_info\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mslave_ok\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1465\u001b[0m             \u001b[0;32mexcept\u001b[0m \u001b[0mServerSelectionTimeoutError\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1466\u001b[0m                 \u001b[0;32mif\u001b[0m \u001b[0mretrying\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/lib/python3.7/site-packages/pymongo/mongo_client.py\u001b[0m in \u001b[0;36m_cmd\u001b[0;34m(session, server, sock_info, slave_ok)\u001b[0m\n\u001b[1;32m   1338\u001b[0m                 \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_event_listeners\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1339\u001b[0m                 \u001b[0mexhaust\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1340\u001b[0;31m                 unpack_res)\n\u001b[0m\u001b[1;32m   1341\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1342\u001b[0m         return self._retryable_read(\n",
      "\u001b[0;32m~/anaconda3/lib/python3.7/site-packages/pymongo/server.py\u001b[0m in \u001b[0;36mrun_operation_with_response\u001b[0;34m(self, sock_info, operation, set_slave_okay, listeners, exhaust, unpack_res)\u001b[0m\n\u001b[1;32m    115\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0msend_message\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    116\u001b[0m                 \u001b[0msock_info\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msend_message\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdata\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmax_doc_size\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 117\u001b[0;31m                 \u001b[0mreply\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0msock_info\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreceive_message\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mrequest_id\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    118\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    119\u001b[0m                 \u001b[0mreply\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0msock_info\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreceive_message\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;32mNone\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/lib/python3.7/site-packages/pymongo/pool.py\u001b[0m in \u001b[0;36mreceive_message\u001b[0;34m(self, request_id)\u001b[0m\n\u001b[1;32m    644\u001b[0m                                    self.max_message_size)\n\u001b[1;32m    645\u001b[0m         \u001b[0;32mexcept\u001b[0m \u001b[0mBaseException\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0merror\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 646\u001b[0;31m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_raise_connection_failure\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0merror\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    647\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    648\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m_raise_if_not_writable\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0munacknowledged\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/lib/python3.7/site-packages/pymongo/pool.py\u001b[0m in \u001b[0;36mreceive_message\u001b[0;34m(self, request_id)\u001b[0m\n\u001b[1;32m    642\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    643\u001b[0m             return receive_message(self.sock, request_id,\n\u001b[0;32m--> 644\u001b[0;31m                                    self.max_message_size)\n\u001b[0m\u001b[1;32m    645\u001b[0m         \u001b[0;32mexcept\u001b[0m \u001b[0mBaseException\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0merror\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    646\u001b[0m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_raise_connection_failure\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0merror\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/lib/python3.7/site-packages/pymongo/network.py\u001b[0m in \u001b[0;36mreceive_message\u001b[0;34m(sock, request_id, max_message_size)\u001b[0m\n\u001b[1;32m    212\u001b[0m             _receive_data_on_socket(sock, length - 25), compressor_id)\n\u001b[1;32m    213\u001b[0m     \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 214\u001b[0;31m         \u001b[0mdata\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0m_receive_data_on_socket\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msock\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mlength\u001b[0m \u001b[0;34m-\u001b[0m \u001b[0;36m16\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    215\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    216\u001b[0m     \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/lib/python3.7/site-packages/pymongo/network.py\u001b[0m in \u001b[0;36m_receive_data_on_socket\u001b[0;34m(sock, length)\u001b[0m\n\u001b[1;32m    253\u001b[0m         \u001b[0;32mwhile\u001b[0m \u001b[0mbytes_read\u001b[0m \u001b[0;34m<\u001b[0m \u001b[0mlength\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    254\u001b[0m             \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 255\u001b[0;31m                 \u001b[0mchunk_length\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0msock\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrecv_into\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmv\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mbytes_read\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    256\u001b[0m             \u001b[0;32mexcept\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mIOError\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mOSError\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0mexc\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    257\u001b[0m                 \u001b[0;32mif\u001b[0m \u001b[0m_errno_from_exception\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mexc\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0merrno\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mEINTR\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "'21_hydroxylase_deficiency'\n",
    "\n",
    "for doc in table.find():\n",
    "    if len(list(\n",
    "        filter(lambda match: match['name'] == '21_hydroxylase_deficiency', doc['synonyms']))) > 1:\n",
    "        print(doc['pref_name'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 138,
   "metadata": {},
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "unhashable type: 'dict'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-138-a965768d0752>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mtable\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcreateIndex\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m{\u001b[0m\u001b[0msynonyms\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0;34m\"text\"\u001b[0m\u001b[0;34m}\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m: unhashable type: 'dict'"
     ]
    }
   ],
   "source": [
    "table.createIndex({synonyms: \"text\"})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "metadata": {},
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "unhashable type: 'dict'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-141-eb8adabe7989>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mdb\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdisease_synonyms\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcreateIndex\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m{\u001b[0m\u001b[0;34m{\u001b[0m\u001b[0msynonyms\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0;34m\"text\"\u001b[0m\u001b[0;34m}\u001b[0m\u001b[0;34m}\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m: unhashable type: 'dict'"
     ]
    }
   ],
   "source": [
    "db.disease_synonyms.createIndex({{synonyms: \"text\"}})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "None\n"
     ]
    }
   ],
   "source": [
    "print(table.find_one({'synonyms':'21_hydroxylase_deficiency'}))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'name': '4_dienoyl_coa_reductase_deficiency_2',\n",
       "  'token': '4_dienoyl_coa_reductase_deficiency_2',\n",
       "  'type': ['synonym'],\n",
       "  'is_synonym': True,\n",
       "  'sources': {'UNIPROT_DISEASE': {'93e19430-7449-4460-a177-c702ec6da40a': ['PREF_NAME']},\n",
       "   'GARD': {'1e0e4587-f638-455a-b3aa-d99366ee3ffb': ['PREF_NAME']},\n",
       "   'ORPHANET': {'ac109ed0-ac25-4935-a3c6-458359bcc828': ['SYNONYM']}},\n",
       "  'original_tokens': [],\n",
       "  'normalized_variants': ['4_dienoyl_coa_reductase_deficiency_2'],\n",
       "  'attributes': {'occurrence_count': 0,\n",
       "   'pubmed_document_count': 0,\n",
       "   'document_count': 0,\n",
       "   'ctrials_document_count': 0,\n",
       "   'uspto_document_count': 0,\n",
       "   'pubmed_concurrence_docs': 0,\n",
       "   'concurrence_document_count': 0,\n",
       "   'p(e|t)_pubmed': 0,\n",
       "   'p(t|e)_pubmed': 0,\n",
       "   'ctrials_concurrence_docs': 0,\n",
       "   'p(e|t)_ctrials': 0,\n",
       "   'p(t|e)_ctrials': 0,\n",
       "   'uspto_concurrence_docs': 0,\n",
       "   'p(e|t)_uspto': 0,\n",
       "   'p(t|e)_uspto': 0,\n",
       "   'source_count': 3,\n",
       "   'concurrence_ontology_count': 3,\n",
       "   'p(t|e)_ontologies': 0.75,\n",
       "   'calculated_syn_weight': 0,\n",
       "   'good_synonym': True,\n",
       "   'logic': 'DUAL_PHRASE_VARIANT',\n",
       "   'p(t|e)_corpus': 0,\n",
       "   'p(e|t)_corpus': 0,\n",
       "   'relevance_to_entity': 0}},\n",
       " {'name': '2_4_dienoyl_coa_reductase_deficiency',\n",
       "  'token': '2_4_dienoyl_coa_reductase_deficiency',\n",
       "  'type': ['synonym', 'pref_name'],\n",
       "  'is_synonym': True,\n",
       "  'sources': {'UNIPROT_DISEASE': {'93e19430-7449-4460-a177-c702ec6da40a': ['PREF_NAME']},\n",
       "   'GARD': {'1e0e4587-f638-455a-b3aa-d99366ee3ffb': ['PREF_NAME']},\n",
       "   'ORPHANET': {'ac109ed0-ac25-4935-a3c6-458359bcc828': ['SYNONYM']},\n",
       "   'UMLS': {'69cd80b8-612f-4b6b-8719-5d7c11e68197': ['PREF_NAME']}},\n",
       "  'original_tokens': [],\n",
       "  'normalized_variants': ['2_4_dienoyl_coa_reductase_deficiency'],\n",
       "  'attributes': {'occurrence_count': 66,\n",
       "   'pubmed_document_count': 2,\n",
       "   'document_count': 2,\n",
       "   'ctrials_document_count': 0,\n",
       "   'uspto_document_count': 0,\n",
       "   'pubmed_concurrence_docs': 2,\n",
       "   'concurrence_document_count': 2,\n",
       "   'p(e|t)_pubmed': 1,\n",
       "   'p(t|e)_pubmed': 1,\n",
       "   'ctrials_concurrence_docs': 0,\n",
       "   'p(e|t)_ctrials': 0,\n",
       "   'p(t|e)_ctrials': 0,\n",
       "   'uspto_concurrence_docs': 0,\n",
       "   'p(e|t)_uspto': 0,\n",
       "   'p(t|e)_uspto': 0,\n",
       "   'source_count': 4,\n",
       "   'concurrence_ontology_count': 4,\n",
       "   'p(t|e)_ontologies': 1,\n",
       "   'calculated_syn_weight': 1,\n",
       "   'good_synonym': True,\n",
       "   'logic': 'PREF_TOKEN',\n",
       "   'p(t|e)_corpus': 0,\n",
       "   'p(e|t)_corpus': 0,\n",
       "   'relevance_to_entity': 1}},\n",
       " {'name': 'dienoyl_coa_reductase_deficiency',\n",
       "  'token': 'dienoyl_coa_reductase_deficiency',\n",
       "  'type': ['synonym'],\n",
       "  'is_synonym': True,\n",
       "  'sources': {'GARD': {'1e0e4587-f638-455a-b3aa-d99366ee3ffb': ['SYNONYM']}},\n",
       "  'original_tokens': [],\n",
       "  'normalized_variants': ['dienoyl_coa_reductase_deficiency'],\n",
       "  'attributes': {'occurrence_count': 647,\n",
       "   'pubmed_document_count': 10,\n",
       "   'document_count': 36,\n",
       "   'ctrials_document_count': 0,\n",
       "   'uspto_document_count': 26,\n",
       "   'pubmed_concurrence_docs': 0,\n",
       "   'concurrence_document_count': 0,\n",
       "   'p(e|t)_pubmed': 0,\n",
       "   'p(t|e)_pubmed': 0,\n",
       "   'ctrials_concurrence_docs': 0,\n",
       "   'p(e|t)_ctrials': 0,\n",
       "   'p(t|e)_ctrials': 0,\n",
       "   'uspto_concurrence_docs': 0,\n",
       "   'p(e|t)_uspto': 0,\n",
       "   'p(t|e)_uspto': 0,\n",
       "   'source_count': 1,\n",
       "   'concurrence_ontology_count': 1,\n",
       "   'p(t|e)_ontologies': 0.25,\n",
       "   'calculated_syn_weight': 0,\n",
       "   'good_synonym': False,\n",
       "   'logic': 'NEW_ENTITY_SCORE',\n",
       "   'p(t|e)_corpus': 0,\n",
       "   'p(e|t)_corpus': 0,\n",
       "   'relevance_to_entity': 0}},\n",
       " {'name': '4_alpha_dienoyl_coa_reductase_deficiency_2',\n",
       "  'token': '4_alpha_dienoyl_coa_reductase_deficiency_2',\n",
       "  'type': ['synonym'],\n",
       "  'is_synonym': True,\n",
       "  'sources': {'GARD': {'1e0e4587-f638-455a-b3aa-d99366ee3ffb': ['SYNONYM']}},\n",
       "  'original_tokens': [],\n",
       "  'normalized_variants': ['4_alpha_dienoyl_coa_reductase_deficiency_2'],\n",
       "  'attributes': {'occurrence_count': 0,\n",
       "   'pubmed_document_count': 0,\n",
       "   'document_count': 0,\n",
       "   'ctrials_document_count': 0,\n",
       "   'uspto_document_count': 0,\n",
       "   'pubmed_concurrence_docs': 0,\n",
       "   'concurrence_document_count': 0,\n",
       "   'p(e|t)_pubmed': 0,\n",
       "   'p(t|e)_pubmed': 0,\n",
       "   'ctrials_concurrence_docs': 0,\n",
       "   'p(e|t)_ctrials': 0,\n",
       "   'p(t|e)_ctrials': 0,\n",
       "   'uspto_concurrence_docs': 0,\n",
       "   'p(e|t)_uspto': 0,\n",
       "   'p(t|e)_uspto': 0,\n",
       "   'source_count': 1,\n",
       "   'concurrence_ontology_count': 1,\n",
       "   'p(t|e)_ontologies': 0.25,\n",
       "   'calculated_syn_weight': 0,\n",
       "   'good_synonym': False,\n",
       "   'logic': 'NEW_ENTITY_SCORE',\n",
       "   'p(t|e)_corpus': 0,\n",
       "   'p(e|t)_corpus': 0,\n",
       "   'relevance_to_entity': 0}},\n",
       " {'name': '2_4_alpha_dienoyl_coa_reductase_deficiency',\n",
       "  'token': '2_4_alpha_dienoyl_coa_reductase_deficiency',\n",
       "  'type': ['synonym'],\n",
       "  'is_synonym': True,\n",
       "  'sources': {'GARD': {'1e0e4587-f638-455a-b3aa-d99366ee3ffb': ['SYNONYM']}},\n",
       "  'original_tokens': [],\n",
       "  'normalized_variants': ['2_4_alpha_dienoyl_coa_reductase_deficiency'],\n",
       "  'attributes': {'occurrence_count': 0,\n",
       "   'pubmed_document_count': 0,\n",
       "   'document_count': 0,\n",
       "   'ctrials_document_count': 0,\n",
       "   'uspto_document_count': 0,\n",
       "   'pubmed_concurrence_docs': 0,\n",
       "   'concurrence_document_count': 0,\n",
       "   'p(e|t)_pubmed': 0,\n",
       "   'p(t|e)_pubmed': 0,\n",
       "   'ctrials_concurrence_docs': 0,\n",
       "   'p(e|t)_ctrials': 0,\n",
       "   'p(t|e)_ctrials': 0,\n",
       "   'uspto_concurrence_docs': 0,\n",
       "   'p(e|t)_uspto': 0,\n",
       "   'p(t|e)_uspto': 0,\n",
       "   'source_count': 1,\n",
       "   'concurrence_ontology_count': 1,\n",
       "   'p(t|e)_ontologies': 0.25,\n",
       "   'calculated_syn_weight': 0,\n",
       "   'good_synonym': False,\n",
       "   'logic': 'NEW_ENTITY_SCORE',\n",
       "   'p(t|e)_corpus': 0,\n",
       "   'p(e|t)_corpus': 0,\n",
       "   'relevance_to_entity': 0}},\n",
       " {'name': 'progressive_encephalopathy_with_leukodystrophy_due_to_decr_deficiency',\n",
       "  'token': 'progressive_encephalopathy_with_leukodystrophy_due_to_decr_deficiency',\n",
       "  'type': ['synonym'],\n",
       "  'is_synonym': True,\n",
       "  'sources': {'ORPHANET': {'ac109ed0-ac25-4935-a3c6-458359bcc828': ['PREF_NAME']}},\n",
       "  'original_tokens': [],\n",
       "  'normalized_variants': ['progressive_encephalopathy_with_leukodystrophy_due_to_decr_deficiency'],\n",
       "  'attributes': {'occurrence_count': 18,\n",
       "   'pubmed_document_count': 0,\n",
       "   'document_count': 0,\n",
       "   'ctrials_document_count': 0,\n",
       "   'uspto_document_count': 0,\n",
       "   'pubmed_concurrence_docs': 0,\n",
       "   'concurrence_document_count': 0,\n",
       "   'p(e|t)_pubmed': 0,\n",
       "   'p(t|e)_pubmed': 0,\n",
       "   'ctrials_concurrence_docs': 0,\n",
       "   'p(e|t)_ctrials': 0,\n",
       "   'p(t|e)_ctrials': 0,\n",
       "   'uspto_concurrence_docs': 0,\n",
       "   'p(e|t)_uspto': 0,\n",
       "   'p(t|e)_uspto': 0,\n",
       "   'source_count': 1,\n",
       "   'concurrence_ontology_count': 1,\n",
       "   'p(t|e)_ontologies': 0.25,\n",
       "   'calculated_syn_weight': 0,\n",
       "   'good_synonym': False,\n",
       "   'logic': 'NEW_ENTITY_SCORE',\n",
       "   'p(t|e)_corpus': 0,\n",
       "   'p(e|t)_corpus': 0,\n",
       "   'relevance_to_entity': 0}},\n",
       " {'name': 'decr_deficiency_with_hyperlysinemia',\n",
       "  'token': 'decr_deficiency_with_hyperlysinemia',\n",
       "  'type': ['synonym'],\n",
       "  'is_synonym': True,\n",
       "  'sources': {'ORPHANET': {'ac109ed0-ac25-4935-a3c6-458359bcc828': ['SYNONYM']}},\n",
       "  'original_tokens': [],\n",
       "  'normalized_variants': ['decr_deficiency_with_hyperlysinemia'],\n",
       "  'attributes': {'occurrence_count': 15,\n",
       "   'pubmed_document_count': 1,\n",
       "   'document_count': 1,\n",
       "   'ctrials_document_count': 0,\n",
       "   'uspto_document_count': 0,\n",
       "   'pubmed_concurrence_docs': 0,\n",
       "   'concurrence_document_count': 0,\n",
       "   'p(e|t)_pubmed': 0,\n",
       "   'p(t|e)_pubmed': 0,\n",
       "   'ctrials_concurrence_docs': 0,\n",
       "   'p(e|t)_ctrials': 0,\n",
       "   'p(t|e)_ctrials': 0,\n",
       "   'uspto_concurrence_docs': 0,\n",
       "   'p(e|t)_uspto': 0,\n",
       "   'p(t|e)_uspto': 0,\n",
       "   'source_count': 1,\n",
       "   'concurrence_ontology_count': 1,\n",
       "   'p(t|e)_ontologies': 0.25,\n",
       "   'calculated_syn_weight': 0,\n",
       "   'good_synonym': False,\n",
       "   'logic': 'NEW_ENTITY_SCORE',\n",
       "   'p(t|e)_corpus': 0,\n",
       "   'p(e|t)_corpus': 0,\n",
       "   'relevance_to_entity': 0}},\n",
       " {'name': 'decrd',\n",
       "  'token': 'decrd',\n",
       "  'type': ['synonym'],\n",
       "  'is_synonym': True,\n",
       "  'sources': {'UMLS': {'69cd80b8-612f-4b6b-8719-5d7c11e68197': ['SYNONYM']}},\n",
       "  'original_tokens': [],\n",
       "  'normalized_variants': ['decrd'],\n",
       "  'attributes': {'occurrence_count': 379,\n",
       "   'pubmed_document_count': 2,\n",
       "   'document_count': 3,\n",
       "   'ctrials_document_count': 0,\n",
       "   'uspto_document_count': 1,\n",
       "   'pubmed_concurrence_docs': 0,\n",
       "   'concurrence_document_count': 0,\n",
       "   'p(e|t)_pubmed': 0,\n",
       "   'p(t|e)_pubmed': 0,\n",
       "   'ctrials_concurrence_docs': 0,\n",
       "   'p(e|t)_ctrials': 0,\n",
       "   'p(t|e)_ctrials': 0,\n",
       "   'uspto_concurrence_docs': 0,\n",
       "   'p(e|t)_uspto': 0,\n",
       "   'p(t|e)_uspto': 0,\n",
       "   'source_count': 1,\n",
       "   'concurrence_ontology_count': 1,\n",
       "   'p(t|e)_ontologies': 0.25,\n",
       "   'relevance_across_entities': 0.5,\n",
       "   'calculated_syn_weight': 0,\n",
       "   'good_synonym': False,\n",
       "   'logic': 'NEW_ENTITY_SCORE',\n",
       "   'p(t|e)_corpus': 0,\n",
       "   'p(e|t)_corpus': 0,\n",
       "   'relevance_to_entity': 0}}]"
      ]
     },
     "execution_count": 144,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "first_instance['synonyms']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 146,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 146,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(table.synonyms.find({'name.decrd': {'$exists': 1}}))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Collection(Database(MongoClient(host=['mongo.servers.nferx.com:27017'], document_class=dict, tz_aware=False, connect=True), 'synonyms_db'), 'disease_synonyms.pref_name')"
      ]
     },
     "execution_count": 156,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "table['pref_name']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
