{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from functools import reduce\n",
    "from concurrent.futures import ThreadPoolExecutor\n",
    "import requests\n",
    "from urllib.parse import urlencode\n",
    "import json\n",
    "import requests\n",
    "import pandas\n",
    "import os\n",
    "from pandas.io.json import json_normalize\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "os.environ['Nfer_Access_Key'] = \"praveen@nference.net\"\n",
    "os.environ['Nfer_Secret_Key'] = \"81d4782ff7cda46fdbd1c2b758af0759\"\n",
    "\n",
    "PRESTAGING_COOKIE = {\n",
    "\"Cookie\":\"csrftoken=wLgolDsSbcb3Py2xB8xuiRDmwezQ9Y9o; sessionid=jqab5fdwqmrxrnc5u0cli8xqxerep2p2\"}\n",
    "\n",
    "headers = {}\n",
    "headers[\"cookie\"] = \"csrftoken=wLgolDsSbcb3Py2xB8xuiRDmwezQ9Y9o; sessionid=jqab5fdwqmrxrnc5u0cli8xqxerep2p2\" "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 295,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_enrichment(hgnc_id,ensemble_id,lowrange,highrange,entity):\n",
    "    url = \"https://preview.nferx.com/dv/201912/rnaapi2/entity_distribution?\"\n",
    "    \n",
    "    params = {\n",
    "        \"entity_type\":entity,\n",
    "        \"range_a_min\":lowrange[0],\n",
    "        \"range_a_max\":lowrange[1],\n",
    "        \"range_b_min\":highrange[0],\n",
    "        \"range_b_max\":highrange[1],\n",
    "        \"enriched_group\": \"B\",\n",
    "        \"exclude_zero\": True,\n",
    "        \"token_extraction\": \"ngram\",\n",
    "        \"single_cell_filter\":\"not_single_cell\",\n",
    "        \"cell_line_filter\":\"not_cell_line\",\n",
    "        \"exclude_zero\":True,\n",
    "        \"gene_id\":ensemble_id,\n",
    "        \"hgnc_id\":hgnc_id\n",
    "    }\n",
    "\n",
    "    x = requests.get(url, params=params, headers = headers)\n",
    "    required_table = json_normalize(x.json()['result']['term_features'])\n",
    "    return(required_table)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_IL1A_disease_enriched = get_enrichment(\"il1a\",\"ensg00000115008\",[0,0.07],[0.0,14],\"root-diseases\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>neg_log_p_bh_adjusted_p</th>\n",
       "      <th>term</th>\n",
       "      <th>sub_entity_type</th>\n",
       "      <th>range_a_count</th>\n",
       "      <th>range_b_count</th>\n",
       "      <th>all_studies_count</th>\n",
       "      <th>gtex_studies_count</th>\n",
       "      <th>tcga_studies_count</th>\n",
       "      <th>signal_token</th>\n",
       "      <th>annotation</th>\n",
       "      <th>ls</th>\n",
       "      <th>gs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>mitral_valve_regurgitation</td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>mitral_valve_regurgitation</td>\n",
       "      <td>[]</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>neurogenic_bladder</td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>neurogenic_bladder</td>\n",
       "      <td>[]</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>sccb</td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>sccb</td>\n",
       "      <td>[]</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>2.789372</td>\n",
       "      <td>h5n1</td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "      <td>42</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>h5n1</td>\n",
       "      <td>[]</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>serous_carcinoma</td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>serous_carcinoma</td>\n",
       "      <td>[]</td>\n",
       "      <td>0.04</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>356</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>monosomy_21</td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>monosomy_21</td>\n",
       "      <td>[]</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>357</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>ppcd</td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>ppcd</td>\n",
       "      <td>[]</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>358</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>anaplastic_oligodendroglioma</td>\n",
       "      <td>disease</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>anaplastic_oligodendroglioma</td>\n",
       "      <td>[]</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>359</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>ataxia_telangiectasia</td>\n",
       "      <td>disease</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>ataxia_telangiectasia</td>\n",
       "      <td>[]</td>\n",
       "      <td>0.68</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>360</td>\n",
       "      <td>49.419930</td>\n",
       "      <td>breast_cancer</td>\n",
       "      <td>disease</td>\n",
       "      <td>169</td>\n",
       "      <td>1974</td>\n",
       "      <td>36</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>breast_cancer</td>\n",
       "      <td>[]</td>\n",
       "      <td>4.17</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>361 rows Ã— 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     neg_log_p_bh_adjusted_p                          term sub_entity_type  \\\n",
       "0                   0.000000    mitral_valve_regurgitation                   \n",
       "1                   0.000000            neurogenic_bladder                   \n",
       "2                   0.000000                          sccb                   \n",
       "3                   2.789372                          h5n1                   \n",
       "4                   0.000000              serous_carcinoma                   \n",
       "..                       ...                           ...             ...   \n",
       "356                 0.000000                   monosomy_21                   \n",
       "357                 0.000000                          ppcd                   \n",
       "358                 0.000000  anaplastic_oligodendroglioma         disease   \n",
       "359                 0.000000         ataxia_telangiectasia         disease   \n",
       "360                49.419930                 breast_cancer         disease   \n",
       "\n",
       "     range_a_count  range_b_count  all_studies_count  gtex_studies_count  \\\n",
       "0                0              1                  1                   0   \n",
       "1                0              1                  1                   0   \n",
       "2                0              7                  1                   0   \n",
       "3                0             42                  4                   0   \n",
       "4                0              4                  2                   0   \n",
       "..             ...            ...                ...                 ...   \n",
       "356              0              2                  1                   0   \n",
       "357              0              1                  1                   0   \n",
       "358              0              1                  1                   0   \n",
       "359              0              6                  1                   0   \n",
       "360            169           1974                 36                   0   \n",
       "\n",
       "     tcga_studies_count                  signal_token annotation    ls   gs  \n",
       "0                     0    mitral_valve_regurgitation         []  0.00  0.0  \n",
       "1                     0            neurogenic_bladder         []  0.00  0.0  \n",
       "2                     0                          sccb         []  0.00  0.0  \n",
       "3                     0                          h5n1         []  0.00  0.0  \n",
       "4                     0              serous_carcinoma         []  0.04  0.0  \n",
       "..                  ...                           ...        ...   ...  ...  \n",
       "356                   0                   monosomy_21         []  0.00  0.0  \n",
       "357                   0                          ppcd         []  0.00  0.0  \n",
       "358                   0  anaplastic_oligodendroglioma         []  0.00  0.0  \n",
       "359                   0         ataxia_telangiectasia         []  0.68  0.0  \n",
       "360                   0                 breast_cancer         []  4.17  0.0  \n",
       "\n",
       "[361 rows x 12 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_IL1A_disease_enriched"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_IL1A_disease_enriched.to_csv('/Users/praveenanand/Documents/Apps/Bermekimab_LabelExpansen_Janssen_May2020/Deliverable2_again/Data/IL1A_RNAseq_disease_enrichment_allrange.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 300,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_IL1A_disease_enriched_updated = get_enrichment(\"il1a\",\"ensg00000115008\",[0,0.06],[3.68,14],\"root-diseases\")\n",
    "all_IL1A_disease_enriched_updated.to_csv('/Users/praveenanand/Documents/Apps/Bermekimab_LabelExpansen_Janssen_May2020/Deliverable2_again/Data/IL1A_RNAseq_disease_enrichment_allrange_updated.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 299,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>neg_log_p_bh_adjusted_p</th>\n",
       "      <th>term</th>\n",
       "      <th>sub_entity_type</th>\n",
       "      <th>range_a_count</th>\n",
       "      <th>range_b_count</th>\n",
       "      <th>all_studies_count</th>\n",
       "      <th>gtex_studies_count</th>\n",
       "      <th>tcga_studies_count</th>\n",
       "      <th>signal_token</th>\n",
       "      <th>annotation</th>\n",
       "      <th>ls</th>\n",
       "      <th>gs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>45</td>\n",
       "      <td>32.802905</td>\n",
       "      <td>lupus</td>\n",
       "      <td></td>\n",
       "      <td>28</td>\n",
       "      <td>80</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>lupus</td>\n",
       "      <td>[]</td>\n",
       "      <td>3.35</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    neg_log_p_bh_adjusted_p   term sub_entity_type  range_a_count  \\\n",
       "45                32.802905  lupus                             28   \n",
       "\n",
       "    range_b_count  all_studies_count  gtex_studies_count  tcga_studies_count  \\\n",
       "45             80                  1                   0                   0   \n",
       "\n",
       "   signal_token annotation    ls   gs  \n",
       "45        lupus         []  3.35  0.0  "
      ]
     },
     "execution_count": 299,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_IL1A_disease_enriched_updated = get_enrichment(\"il1a\",\"ensg00000115008\",[0,0.06],[3.68,14],\"root-diseases\")\n",
    "all_IL1A_disease_enriched_updated.loc[all_IL1A_disease_enriched_updated['term'] == 'lupus']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_IL1A_disease_enriched_updated = get_enrichment(\"il1a\",\"ensg00000115008\",[0,3.859],[3.86,14],\"root-diseases\")\n",
    "all_IL1A_disease_enriched.to_csv('/Users/praveenanand/Documents/Apps/Bermekimab_LabelExpansen_Janssen_May2020/Deliverable2_again/Data/IL1A_RNAseq_disease_enrichment_allrange_updated_allrange.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_IL1A_disease_enriched_updated"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_enrichment(hgnc_id,ensemble_id,lowrange,highrange,entity):\n",
    "    url = \"https://preview.nferx.com/dv/201912/rnaapi2/entity_distribution?\"\n",
    "    params = {\n",
    "        \"entity_type\":entity,\n",
    "        \"range_a_min\":lowrange[0],\n",
    "        \"range_a_max\":lowrange[1],\n",
    "        \"range_b_min\":highrange[0],\n",
    "        \"range_b_max\":highrange[1],\n",
    "        \"enriched_group\": \"B\",\n",
    "        \"exclude_zero\": True,\n",
    "        \"token_extraction\": \"ngram\",\n",
    "        \"single_cell_filter\":\"not_single_cell\",\n",
    "        \"cell_line_filter\":\"not_cell_line\",\n",
    "        \"exclude_zero\":True,\n",
    "        \"gene_id\":ensemble_id,\n",
    "        \"hgnc_id\":hgnc_id\n",
    "    }\n",
    "    x = requests.get(url, params=params, headers = headers)\n",
    "    required_table = json_normalize(x.json()['result']['term_features'])\n",
    "    return(required_table)\n",
    "all_IL1A_disease_enriched_actual = get_enrichment(\"il1a\",\"ensg00000115008\",[0,0.06],[3.68,14],\"root-diseases\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>neg_log_p_bh_adjusted_p</th>\n",
       "      <th>term</th>\n",
       "      <th>sub_entity_type</th>\n",
       "      <th>range_a_count</th>\n",
       "      <th>range_b_count</th>\n",
       "      <th>all_studies_count</th>\n",
       "      <th>gtex_studies_count</th>\n",
       "      <th>tcga_studies_count</th>\n",
       "      <th>signal_token</th>\n",
       "      <th>annotation</th>\n",
       "      <th>ls</th>\n",
       "      <th>gs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>92</td>\n",
       "      <td>39.54073</td>\n",
       "      <td>squamous_cell_carcinoma</td>\n",
       "      <td>disease</td>\n",
       "      <td>15</td>\n",
       "      <td>712</td>\n",
       "      <td>19</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>squamous_cell_carcinoma</td>\n",
       "      <td>[tcga_evidence]</td>\n",
       "      <td>2.66</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    neg_log_p_bh_adjusted_p                     term sub_entity_type  \\\n",
       "92                 39.54073  squamous_cell_carcinoma         disease   \n",
       "\n",
       "    range_a_count  range_b_count  all_studies_count  gtex_studies_count  \\\n",
       "92             15            712                 19                   0   \n",
       "\n",
       "    tcga_studies_count             signal_token       annotation    ls   gs  \n",
       "92                   4  squamous_cell_carcinoma  [tcga_evidence]  2.66  0.0  "
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_IL1A_disease_enriched_all = get_enrichment(\"il1a\",\"ensg00000115008\",[0,0.06],[0,14],\"root-diseases\")\n",
    "all_IL1A_disease_enriched_all.loc[all_IL1A_disease_enriched_all['term'] == 'squamous_cell_carcinoma']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>neg_log_p_bh_adjusted_p</th>\n",
       "      <th>term</th>\n",
       "      <th>sub_entity_type</th>\n",
       "      <th>range_a_count</th>\n",
       "      <th>range_b_count</th>\n",
       "      <th>all_studies_count</th>\n",
       "      <th>gtex_studies_count</th>\n",
       "      <th>tcga_studies_count</th>\n",
       "      <th>signal_token</th>\n",
       "      <th>annotation</th>\n",
       "      <th>ls</th>\n",
       "      <th>gs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>25</td>\n",
       "      <td>73.882236</td>\n",
       "      <td>squamous_cell_carcinoma</td>\n",
       "      <td>disease</td>\n",
       "      <td>15</td>\n",
       "      <td>132</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>squamous_cell_carcinoma</td>\n",
       "      <td>[tcga_evidence]</td>\n",
       "      <td>2.66</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    neg_log_p_bh_adjusted_p                     term sub_entity_type  \\\n",
       "25                73.882236  squamous_cell_carcinoma         disease   \n",
       "\n",
       "    range_a_count  range_b_count  all_studies_count  gtex_studies_count  \\\n",
       "25             15            132                  8                   0   \n",
       "\n",
       "    tcga_studies_count             signal_token       annotation    ls   gs  \n",
       "25                   1  squamous_cell_carcinoma  [tcga_evidence]  2.66  0.0  "
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_IL1A_disease_enriched_actual = get_enrichment(\"il1a\",\"ensg00000115008\",[0,0.06],[3.68,14],\"root-diseases\")\n",
    "all_IL1A_disease_enriched_actual.loc[all_IL1A_disease_enriched_actual['term'] == 'squamous_cell_carcinoma']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>neg_log_p_bh_adjusted_p</th>\n",
       "      <th>term</th>\n",
       "      <th>sub_entity_type</th>\n",
       "      <th>range_a_count</th>\n",
       "      <th>range_b_count</th>\n",
       "      <th>all_studies_count</th>\n",
       "      <th>gtex_studies_count</th>\n",
       "      <th>tcga_studies_count</th>\n",
       "      <th>signal_token</th>\n",
       "      <th>annotation</th>\n",
       "      <th>ls</th>\n",
       "      <th>gs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>29</td>\n",
       "      <td>42.822632</td>\n",
       "      <td>squamous_cell_carcinoma</td>\n",
       "      <td>disease</td>\n",
       "      <td>15</td>\n",
       "      <td>565</td>\n",
       "      <td>19</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>squamous_cell_carcinoma</td>\n",
       "      <td>[tcga_evidence]</td>\n",
       "      <td>2.66</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    neg_log_p_bh_adjusted_p                     term sub_entity_type  \\\n",
       "29                42.822632  squamous_cell_carcinoma         disease   \n",
       "\n",
       "    range_a_count  range_b_count  all_studies_count  gtex_studies_count  \\\n",
       "29             15            565                 19                   0   \n",
       "\n",
       "    tcga_studies_count             signal_token       annotation    ls   gs  \n",
       "29                   4  squamous_cell_carcinoma  [tcga_evidence]  2.66  0.0  "
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_IL1A_disease_enriched_med = get_enrichment(\"il1a\",\"ensg00000115008\",[0,0.06],[0.061,3.6799],\"root-diseases\")\n",
    "all_IL1A_disease_enriched_med.loc[all_IL1A_disease_enriched_med['term'] == 'squamous_cell_carcinoma']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>neg_log_p_bh_adjusted_p</th>\n",
       "      <th>term</th>\n",
       "      <th>sub_entity_type</th>\n",
       "      <th>range_a_count</th>\n",
       "      <th>range_b_count</th>\n",
       "      <th>all_studies_count</th>\n",
       "      <th>gtex_studies_count</th>\n",
       "      <th>tcga_studies_count</th>\n",
       "      <th>signal_token</th>\n",
       "      <th>annotation</th>\n",
       "      <th>ls</th>\n",
       "      <th>gs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>378</td>\n",
       "      <td>0</td>\n",
       "      <td>squamous_cell_carcinoma</td>\n",
       "      <td>disease</td>\n",
       "      <td>15</td>\n",
       "      <td>15</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>squamous_cell_carcinoma</td>\n",
       "      <td>[tcga_evidence]</td>\n",
       "      <td>2.66</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     neg_log_p_bh_adjusted_p                     term sub_entity_type  \\\n",
       "378                        0  squamous_cell_carcinoma         disease   \n",
       "\n",
       "     range_a_count  range_b_count  all_studies_count  gtex_studies_count  \\\n",
       "378             15             15                  7                   0   \n",
       "\n",
       "     tcga_studies_count             signal_token       annotation    ls   gs  \n",
       "378                   2  squamous_cell_carcinoma  [tcga_evidence]  2.66  0.0  "
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_IL1A_disease_enriched_low = get_enrichment(\"il1a\",\"ensg00000115008\",[0,0.06],[0,0.06],\"root-diseases\")\n",
    "all_IL1A_disease_enriched_low.loc[all_IL1A_disease_enriched_low['term'] == 'squamous_cell_carcinoma']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'squamous_cell_carcinoma'"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test = all_IL1A_disease_enriched_low.loc[all_IL1A_disease_enriched_low['term'] == 'squamous_cell_carcinoma']\n",
    "test['term'].values[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>neg_log_p_bh_adjusted_p</th>\n",
       "      <th>term</th>\n",
       "      <th>sub_entity_type</th>\n",
       "      <th>range_a_count</th>\n",
       "      <th>range_b_count</th>\n",
       "      <th>all_studies_count</th>\n",
       "      <th>gtex_studies_count</th>\n",
       "      <th>tcga_studies_count</th>\n",
       "      <th>signal_token</th>\n",
       "      <th>annotation</th>\n",
       "      <th>ls</th>\n",
       "      <th>gs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [neg_log_p_bh_adjusted_p, term, sub_entity_type, range_a_count, range_b_count, all_studies_count, gtex_studies_count, tcga_studies_count, signal_token, annotation, ls, gs]\n",
       "Index: []"
      ]
     },
     "execution_count": 86,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_IL1A_disease_enriched_all = get_enrichment(\"il1a\",\"ensg00000115008\",[0,0.06],[0.061,14],\"root-diseases\")\n",
    "all_IL1A_disease_enriched_all.loc[all_IL1A_disease_enriched_all['term'] == 'lupus']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_disease_numbers(diseasename):\n",
    "    all_IL1A_disease_enriched_all = get_enrichment(\"il1a\",\"ensg00000115008\",[0,0.06],[0,14],\"root-diseases\")\n",
    "    required_dataall = all_IL1A_disease_enriched_all.loc[all_IL1A_disease_enriched_all['term'] == diseasename]\n",
    "    \n",
    "    all_IL1A_disease_enriched_actual = get_enrichment(\"il1a\",\"ensg00000115008\",[0,0.06],[3.68,14],\"root-diseases\")\n",
    "    required_datahigh = all_IL1A_disease_enriched_actual.loc[all_IL1A_disease_enriched_actual['term'] == diseasename]\n",
    "    \n",
    "    all_IL1A_disease_enriched_med = get_enrichment(\"il1a\",\"ensg00000115008\",[0,0.06],[0.061,3.6799],\"root-diseases\")\n",
    "    required_datamed = all_IL1A_disease_enriched_med.loc[all_IL1A_disease_enriched_med['term'] == diseasename]\n",
    "    \n",
    "    all_IL1A_disease_enriched_low = get_enrichment(\"il1a\",\"ensg00000115008\",[0,0.00001],[0,0.06],\"root-diseases\")\n",
    "    required_datalow = all_IL1A_disease_enriched_low.loc[all_IL1A_disease_enriched_low['term'] == diseasename]\n",
    "    \n",
    "    #combined_result = str(required_datahigh['neg_log_p_bh_adjusted_p'].values[0]) + \",\" + str(required_datahigh['range_b_count'].values[0]) + \",\" + str(required_datahigh['all_studies_count'].values[0]) + \",\" + str(required_datahigh['range_a_count'].values[0]) + \",\" + str(required_datalow['all_studies_count'].values[0]) + \",\" + str(required_datamed['range_b_count'].values[0]) + \",\" + str(required_datamed['all_studies_count'].values[0]) + \",\" + str(required_dataall['range_b_count'].values[0]) + \",\" + str(required_dataall['all_studies_count'].values[0])\n",
    "    \n",
    "    return(required_datahigh)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'73.88223559773495,132,8,15,7,565,19,712,19'"
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "get_disease_numbers('squamous_cell_carcinoma')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>neg_log_p_bh_adjusted_p</th>\n",
       "      <th>term</th>\n",
       "      <th>sub_entity_type</th>\n",
       "      <th>range_a_count</th>\n",
       "      <th>range_b_count</th>\n",
       "      <th>all_studies_count</th>\n",
       "      <th>gtex_studies_count</th>\n",
       "      <th>tcga_studies_count</th>\n",
       "      <th>signal_token</th>\n",
       "      <th>annotation</th>\n",
       "      <th>ls</th>\n",
       "      <th>gs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [neg_log_p_bh_adjusted_p, term, sub_entity_type, range_a_count, range_b_count, all_studies_count, gtex_studies_count, tcga_studies_count, signal_token, annotation, ls, gs]\n",
       "Index: []"
      ]
     },
     "execution_count": 108,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "get_disease_numbers('lupus')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>neg_log_p_bh_adjusted_p</th>\n",
       "      <th>term</th>\n",
       "      <th>sub_entity_type</th>\n",
       "      <th>range_a_count</th>\n",
       "      <th>range_b_count</th>\n",
       "      <th>all_studies_count</th>\n",
       "      <th>gtex_studies_count</th>\n",
       "      <th>tcga_studies_count</th>\n",
       "      <th>signal_token</th>\n",
       "      <th>annotation</th>\n",
       "      <th>ls</th>\n",
       "      <th>gs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>110</td>\n",
       "      <td>3.923852</td>\n",
       "      <td>acute_on_chronic_liver_failure</td>\n",
       "      <td>disease</td>\n",
       "      <td>0</td>\n",
       "      <td>61</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>acute_on_chronic_liver_failure</td>\n",
       "      <td>[]</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     neg_log_p_bh_adjusted_p                            term sub_entity_type  \\\n",
       "110                 3.923852  acute_on_chronic_liver_failure         disease   \n",
       "\n",
       "     range_a_count  range_b_count  all_studies_count  gtex_studies_count  \\\n",
       "110              0             61                  2                   0   \n",
       "\n",
       "     tcga_studies_count                    signal_token annotation   ls   gs  \n",
       "110                   0  acute_on_chronic_liver_failure         []  0.0  0.0  "
      ]
     },
     "execution_count": 101,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "get_disease_numbers('acute_on_chronic_liver_failure')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>neg_log_p_bh_adjusted_p</th>\n",
       "      <th>term</th>\n",
       "      <th>sub_entity_type</th>\n",
       "      <th>range_a_count</th>\n",
       "      <th>range_b_count</th>\n",
       "      <th>all_studies_count</th>\n",
       "      <th>gtex_studies_count</th>\n",
       "      <th>tcga_studies_count</th>\n",
       "      <th>signal_token</th>\n",
       "      <th>annotation</th>\n",
       "      <th>ls</th>\n",
       "      <th>gs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>59</td>\n",
       "      <td>35.544832</td>\n",
       "      <td>lupus</td>\n",
       "      <td></td>\n",
       "      <td>28</td>\n",
       "      <td>80</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>lupus</td>\n",
       "      <td>[]</td>\n",
       "      <td>3.37</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    neg_log_p_bh_adjusted_p   term sub_entity_type  range_a_count  \\\n",
       "59                35.544832  lupus                             28   \n",
       "\n",
       "    range_b_count  all_studies_count  gtex_studies_count  tcga_studies_count  \\\n",
       "59             80                  1                   0                   0   \n",
       "\n",
       "   signal_token annotation    ls  gs  \n",
       "59        lupus         []  3.37   0  "
      ]
     },
     "execution_count": 113,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "url = \"https://preview.nferx.com/dv/201912/rnaapi2/entity_distribution?\"\n",
    "\n",
    "#https://preview.nferx.com/rnaapi2/entity_distribution?hgnc_id=il1a&gene_id=ensg00000115008&range_a_min=0&range_a_max=0.07&range_b_min=3.86&range_b_max=14&entity_type=root-cells-and-tissues&enriched_group=B&cell_line_filter=not_cell_line&single_cell_filter=not_single_cell&exclude_zero=true&token_extraction=ngram\n",
    "\n",
    "params = {\n",
    "    \"entity_type\":\"root-diseases\",\n",
    "    \"range_a_min\":0,\n",
    "    \"range_a_max\":0.06,\n",
    "    \"range_b_min\":3.6799,\n",
    "    \"range_b_max\":14,\n",
    "    \"enriched_group\": \"B\",\n",
    "    \"exclude_zero\": True,\n",
    "    \"token_extraction\": \"ngram\",\n",
    "    \"single_cell_filter\":\"not_single_cell\",\n",
    "    \"cell_line_filter\":\"all\",\n",
    "    \"exclude_zero\":True,\n",
    "    \"gene_id\":\"ensg00000115008\",\n",
    "    \"hgnc_id\":\"IL1A\"\n",
    "}\n",
    "\n",
    "x = requests.get(url, params=params, headers = headers)\n",
    "all_data = json_normalize(x.json()['result']['term_features'])\n",
    "all_data.loc[all_data['term'] == 'lupus']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>neg_log_p_bh_adjusted_p</th>\n",
       "      <th>term</th>\n",
       "      <th>sub_entity_type</th>\n",
       "      <th>range_a_count</th>\n",
       "      <th>range_b_count</th>\n",
       "      <th>all_studies_count</th>\n",
       "      <th>gtex_studies_count</th>\n",
       "      <th>tcga_studies_count</th>\n",
       "      <th>signal_token</th>\n",
       "      <th>annotation</th>\n",
       "      <th>ls</th>\n",
       "      <th>gs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [neg_log_p_bh_adjusted_p, term, sub_entity_type, range_a_count, range_b_count, all_studies_count, gtex_studies_count, tcga_studies_count, signal_token, annotation, ls, gs]\n",
       "Index: []"
      ]
     },
     "execution_count": 115,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "url = \"https://preview.nferx.com/dv/201912/rnaapi2/entity_distribution?\"\n",
    "\n",
    "#https://preview.nferx.com/rnaapi2/entity_distribution?hgnc_id=il1a&gene_id=ensg00000115008&range_a_min=0&range_a_max=0.07&range_b_min=3.86&range_b_max=14&entity_type=root-cells-and-tissues&enriched_group=B&cell_line_filter=not_cell_line&single_cell_filter=not_single_cell&exclude_zero=true&token_extraction=ngram\n",
    "\n",
    "params = {\n",
    "    \"entity_type\":\"root-diseases\",\n",
    "    \"range_a_min\":0,\n",
    "    \"range_a_max\":0.06,\n",
    "    \"range_b_min\":0.061,\n",
    "    \"range_b_max\":14,\n",
    "    \"enriched_group\": \"B\",\n",
    "    \"exclude_zero\": True,\n",
    "    \"token_extraction\": \"ngram\",\n",
    "    \"single_cell_filter\":\"not_single_cell\",\n",
    "    \"cell_line_filter\":\"all\",\n",
    "    \"exclude_zero\":True,\n",
    "    \"gene_id\":\"ensg00000115008\",\n",
    "    \"hgnc_id\":\"IL1A\"\n",
    "}\n",
    "\n",
    "x = requests.get(url, params=params, headers = headers)\n",
    "all_data = json_normalize(x.json()['result']['term_features'])\n",
    "all_data.loc[all_data['term'] == 'lupus']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>neg_log_p_bh_adjusted_p</th>\n",
       "      <th>term</th>\n",
       "      <th>sub_entity_type</th>\n",
       "      <th>range_a_count</th>\n",
       "      <th>range_b_count</th>\n",
       "      <th>all_studies_count</th>\n",
       "      <th>gtex_studies_count</th>\n",
       "      <th>tcga_studies_count</th>\n",
       "      <th>signal_token</th>\n",
       "      <th>annotation</th>\n",
       "      <th>ls</th>\n",
       "      <th>gs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>45</td>\n",
       "      <td>45.28819</td>\n",
       "      <td>breast_cancer</td>\n",
       "      <td>disease</td>\n",
       "      <td>163</td>\n",
       "      <td>191</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>breast_cancer</td>\n",
       "      <td>[]</td>\n",
       "      <td>4.17</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    neg_log_p_bh_adjusted_p           term sub_entity_type  range_a_count  \\\n",
       "45                 45.28819  breast_cancer         disease            163   \n",
       "\n",
       "    range_b_count  all_studies_count  gtex_studies_count  tcga_studies_count  \\\n",
       "45            191                  3                   0                   0   \n",
       "\n",
       "     signal_token annotation    ls   gs  \n",
       "45  breast_cancer         []  4.17  0.0  "
      ]
     },
     "execution_count": 118,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "get_disease_numbers('breast_cancer')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "https://preview.nferx.com/rnaapi2/term_distribution?gene_id=ensg00000115008&range_a_min=0&range_a_max=0.06&range_b_min=.061&range_b_max=14&entity_type=root-diseases&enriched_group=B&cell_line_filter=not_cell_line&exclude_zero=true&token_extraction=ngram&term=lupus"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>study_id</th>\n",
       "      <th>total_sample_count</th>\n",
       "      <th>study_title</th>\n",
       "      <th>study_type</th>\n",
       "      <th>study_link</th>\n",
       "      <th>samples_a</th>\n",
       "      <th>samples_b</th>\n",
       "      <th>samples_a_cell_line</th>\n",
       "      <th>samples_b_cell_line</th>\n",
       "      <th>samples_a_single_cell</th>\n",
       "      <th>samples_b_single_cell</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>GSE110685</td>\n",
       "      <td>53</td>\n",
       "      <td>Low density granulocytes and their gene signat...</td>\n",
       "      <td>GEO</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi...</td>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>GSE124939</td>\n",
       "      <td>164</td>\n",
       "      <td>Hypersensitive interferon responses in lupus k...</td>\n",
       "      <td>GEO</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi...</td>\n",
       "      <td>0</td>\n",
       "      <td>81</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>GSE139358</td>\n",
       "      <td>41</td>\n",
       "      <td>Transcriptomic, epigenetic and functional anal...</td>\n",
       "      <td>GEO</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi...</td>\n",
       "      <td>0</td>\n",
       "      <td>20</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>GSE86884</td>\n",
       "      <td>96</td>\n",
       "      <td>Differentially Expressed Gene Transcripts Usin...</td>\n",
       "      <td>GEO</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi...</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    study_id  total_sample_count  \\\n",
       "0  GSE110685                  53   \n",
       "1  GSE124939                 164   \n",
       "2  GSE139358                  41   \n",
       "3   GSE86884                  96   \n",
       "\n",
       "                                         study_title study_type  \\\n",
       "0  Low density granulocytes and their gene signat...        GEO   \n",
       "1  Hypersensitive interferon responses in lupus k...        GEO   \n",
       "2  Transcriptomic, epigenetic and functional anal...        GEO   \n",
       "3  Differentially Expressed Gene Transcripts Usin...        GEO   \n",
       "\n",
       "                                          study_link  samples_a  samples_b  \\\n",
       "0  https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi...          0         11   \n",
       "1  https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi...          0         81   \n",
       "2  https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi...          0         20   \n",
       "3  https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi...          0          2   \n",
       "\n",
       "   samples_a_cell_line  samples_b_cell_line  samples_a_single_cell  \\\n",
       "0                    0                    0                      0   \n",
       "1                    0                    0                      0   \n",
       "2                    0                    0                      0   \n",
       "3                    0                    0                      0   \n",
       "\n",
       "   samples_b_single_cell  \n",
       "0                      0  \n",
       "1                      0  \n",
       "2                      0  \n",
       "3                      0  "
      ]
     },
     "execution_count": 124,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "url = \"https://preview.nferx.com/dv/201912/rnaapi2/term_distribution?\"\n",
    "\n",
    "#https://preview.nferx.com/rnaapi2/entity_distribution?hgnc_id=il1a&gene_id=ensg00000115008&range_a_min=0&range_a_max=0.07&range_b_min=3.86&range_b_max=14&entity_type=root-cells-and-tissues&enriched_group=B&cell_line_filter=not_cell_line&single_cell_filter=not_single_cell&exclude_zero=true&token_extraction=ngram\n",
    "\n",
    "params = {\n",
    "    \"entity_type\":\"root-diseases\",\n",
    "    \"range_a_min\":0,\n",
    "    \"range_a_max\":0.0001,\n",
    "    \"range_b_min\":0.001,\n",
    "    \"range_b_max\":14,\n",
    "    \"enriched_group\": \"B\",\n",
    "    \"exclude_zero\": True,\n",
    "    \"token_extraction\": \"ngram\",\n",
    "    \"single_cell_filter\":\"not_single_cell\",\n",
    "    \"cell_line_filter\":\"all\",\n",
    "    \"exclude_zero\":True,\n",
    "    \"gene_id\":\"ensg00000115008\",\n",
    "    \"hgnc_id\":\"IL1A\",\n",
    "    \"term\" : \"lupus\"\n",
    "}\n",
    "\n",
    "x = requests.get(url, params=params, headers = headers)\n",
    "all_data = json_normalize(x.json()['result']['term_features']['studies_found'])\n",
    "all_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>study_id</th>\n",
       "      <th>total_sample_count</th>\n",
       "      <th>study_title</th>\n",
       "      <th>study_type</th>\n",
       "      <th>study_link</th>\n",
       "      <th>samples_a</th>\n",
       "      <th>samples_b</th>\n",
       "      <th>samples_a_cell_line</th>\n",
       "      <th>samples_b_cell_line</th>\n",
       "      <th>samples_a_single_cell</th>\n",
       "      <th>samples_b_single_cell</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>GSE124939</td>\n",
       "      <td>164</td>\n",
       "      <td>Hypersensitive interferon responses in lupus k...</td>\n",
       "      <td>GEO</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi...</td>\n",
       "      <td>0</td>\n",
       "      <td>80</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    study_id  total_sample_count  \\\n",
       "0  GSE124939                 164   \n",
       "\n",
       "                                         study_title study_type  \\\n",
       "0  Hypersensitive interferon responses in lupus k...        GEO   \n",
       "\n",
       "                                          study_link  samples_a  samples_b  \\\n",
       "0  https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi...          0         80   \n",
       "\n",
       "   samples_a_cell_line  samples_b_cell_line  samples_a_single_cell  \\\n",
       "0                    0                    0                      0   \n",
       "\n",
       "   samples_b_single_cell  \n",
       "0                      0  "
      ]
     },
     "execution_count": 122,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "url = \"https://preview.nferx.com/dv/201912/rnaapi2/term_distribution?\"\n",
    "\n",
    "#https://preview.nferx.com/rnaapi2/entity_distribution?hgnc_id=il1a&gene_id=ensg00000115008&range_a_min=0&range_a_max=0.07&range_b_min=3.86&range_b_max=14&entity_type=root-cells-and-tissues&enriched_group=B&cell_line_filter=not_cell_line&single_cell_filter=not_single_cell&exclude_zero=true&token_extraction=ngram\n",
    "\n",
    "params = {\n",
    "    \"entity_type\":\"root-diseases\",\n",
    "    \"range_a_min\":0,\n",
    "    \"range_a_max\":0.06,\n",
    "    \"range_b_min\":3.68,\n",
    "    \"range_b_max\":14,\n",
    "    \"enriched_group\": \"B\",\n",
    "    \"exclude_zero\": True,\n",
    "    \"token_extraction\": \"ngram\",\n",
    "    \"single_cell_filter\":\"not_single_cell\",\n",
    "    \"cell_line_filter\":\"all\",\n",
    "    \"exclude_zero\":True,\n",
    "    \"gene_id\":\"ensg00000115008\",\n",
    "    \"hgnc_id\":\"IL1A\",\n",
    "    \"term\" : \"lupus\"\n",
    "}\n",
    "\n",
    "x = requests.get(url, params=params, headers = headers)\n",
    "all_data = json_normalize(x.json()['result']['term_features']['studies_found'])\n",
    "all_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>study_id</th>\n",
       "      <th>total_sample_count</th>\n",
       "      <th>study_title</th>\n",
       "      <th>study_type</th>\n",
       "      <th>study_link</th>\n",
       "      <th>samples_a</th>\n",
       "      <th>samples_b</th>\n",
       "      <th>samples_a_cell_line</th>\n",
       "      <th>samples_b_cell_line</th>\n",
       "      <th>samples_a_single_cell</th>\n",
       "      <th>samples_b_single_cell</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>GSE124939</td>\n",
       "      <td>164</td>\n",
       "      <td>Hypersensitive interferon responses in lupus k...</td>\n",
       "      <td>GEO</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>GSE139358</td>\n",
       "      <td>41</td>\n",
       "      <td>Transcriptomic, epigenetic and functional anal...</td>\n",
       "      <td>GEO</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi...</td>\n",
       "      <td>15</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    study_id  total_sample_count  \\\n",
       "0  GSE124939                 164   \n",
       "1  GSE139358                  41   \n",
       "\n",
       "                                         study_title study_type  \\\n",
       "0  Hypersensitive interferon responses in lupus k...        GEO   \n",
       "1  Transcriptomic, epigenetic and functional anal...        GEO   \n",
       "\n",
       "                                          study_link  samples_a  samples_b  \\\n",
       "0  https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi...          0          1   \n",
       "1  https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi...         15          5   \n",
       "\n",
       "   samples_a_cell_line  samples_b_cell_line  samples_a_single_cell  \\\n",
       "0                    0                    0                      0   \n",
       "1                    0                    0                      0   \n",
       "\n",
       "   samples_b_single_cell  \n",
       "0                      0  \n",
       "1                      0  "
      ]
     },
     "execution_count": 125,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "url = \"https://preview.nferx.com/dv/201912/rnaapi2/term_distribution?\"\n",
    "\n",
    "#https://preview.nferx.com/rnaapi2/entity_distribution?hgnc_id=il1a&gene_id=ensg00000115008&range_a_min=0&range_a_max=0.07&range_b_min=3.86&range_b_max=14&entity_type=root-cells-and-tissues&enriched_group=B&cell_line_filter=not_cell_line&single_cell_filter=not_single_cell&exclude_zero=true&token_extraction=ngram\n",
    "\n",
    "params = {\n",
    "    \"entity_type\":\"root-diseases\",\n",
    "    \"range_a_min\":0,\n",
    "    \"range_a_max\":0.06,\n",
    "    \"range_b_min\":0.061,\n",
    "    \"range_b_max\":3.68,\n",
    "    \"enriched_group\": \"B\",\n",
    "    \"exclude_zero\": True,\n",
    "    \"token_extraction\": \"ngram\",\n",
    "    \"single_cell_filter\":\"not_single_cell\",\n",
    "    \"cell_line_filter\":\"all\",\n",
    "    \"exclude_zero\":True,\n",
    "    \"gene_id\":\"ensg00000115008\",\n",
    "    \"hgnc_id\":\"IL1A\",\n",
    "    \"term\" : \"lupus\"\n",
    "}\n",
    "\n",
    "x = requests.get(url, params=params, headers = headers)\n",
    "all_data = json_normalize(x.json()['result']['term_features']['studies_found'])\n",
    "all_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>study_id</th>\n",
       "      <th>total_sample_count</th>\n",
       "      <th>study_title</th>\n",
       "      <th>study_type</th>\n",
       "      <th>study_link</th>\n",
       "      <th>samples_a</th>\n",
       "      <th>samples_b</th>\n",
       "      <th>samples_a_cell_line</th>\n",
       "      <th>samples_b_cell_line</th>\n",
       "      <th>samples_a_single_cell</th>\n",
       "      <th>samples_b_single_cell</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>GSE110685</td>\n",
       "      <td>53</td>\n",
       "      <td>Low density granulocytes and their gene signat...</td>\n",
       "      <td>GEO</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi...</td>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>GSE139358</td>\n",
       "      <td>41</td>\n",
       "      <td>Transcriptomic, epigenetic and functional anal...</td>\n",
       "      <td>GEO</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi...</td>\n",
       "      <td>0</td>\n",
       "      <td>15</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>GSE86884</td>\n",
       "      <td>96</td>\n",
       "      <td>Differentially Expressed Gene Transcripts Usin...</td>\n",
       "      <td>GEO</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi...</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    study_id  total_sample_count  \\\n",
       "0  GSE110685                  53   \n",
       "1  GSE139358                  41   \n",
       "2   GSE86884                  96   \n",
       "\n",
       "                                         study_title study_type  \\\n",
       "0  Low density granulocytes and their gene signat...        GEO   \n",
       "1  Transcriptomic, epigenetic and functional anal...        GEO   \n",
       "2  Differentially Expressed Gene Transcripts Usin...        GEO   \n",
       "\n",
       "                                          study_link  samples_a  samples_b  \\\n",
       "0  https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi...          0         11   \n",
       "1  https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi...          0         15   \n",
       "2  https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi...          0          2   \n",
       "\n",
       "   samples_a_cell_line  samples_b_cell_line  samples_a_single_cell  \\\n",
       "0                    0                    0                      0   \n",
       "1                    0                    0                      0   \n",
       "2                    0                    0                      0   \n",
       "\n",
       "   samples_b_single_cell  \n",
       "0                      0  \n",
       "1                      0  \n",
       "2                      0  "
      ]
     },
     "execution_count": 127,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "url = \"https://preview.nferx.com/dv/201912/rnaapi2/term_distribution?\"\n",
    "\n",
    "#https://preview.nferx.com/rnaapi2/entity_distribution?hgnc_id=il1a&gene_id=ensg00000115008&range_a_min=0&range_a_max=0.07&range_b_min=3.86&range_b_max=14&entity_type=root-cells-and-tissues&enriched_group=B&cell_line_filter=not_cell_line&single_cell_filter=not_single_cell&exclude_zero=true&token_extraction=ngram\n",
    "\n",
    "params = {\n",
    "    \"entity_type\":\"root-diseases\",\n",
    "    \"range_a_min\":0,\n",
    "    \"range_a_max\":0.0001,\n",
    "    \"range_b_min\":0.0001,\n",
    "    \"range_b_max\":0.06,\n",
    "    \"enriched_group\": \"B\",\n",
    "    \"exclude_zero\": True,\n",
    "    \"token_extraction\": \"ngram\",\n",
    "    \"single_cell_filter\":\"not_single_cell\",\n",
    "    \"cell_line_filter\":\"all\",\n",
    "    \"exclude_zero\":True,\n",
    "    \"gene_id\":\"ensg00000115008\",\n",
    "    \"hgnc_id\":\"IL1A\",\n",
    "    \"term\" : \"lupus\"\n",
    "}\n",
    "\n",
    "x = requests.get(url, params=params, headers = headers)\n",
    "all_data = json_normalize(x.json()['result']['term_features']['studies_found'])\n",
    "all_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>study_id</th>\n",
       "      <th>total_sample_count</th>\n",
       "      <th>study_title</th>\n",
       "      <th>study_type</th>\n",
       "      <th>study_link</th>\n",
       "      <th>samples_a</th>\n",
       "      <th>samples_b</th>\n",
       "      <th>samples_a_cell_line</th>\n",
       "      <th>samples_b_cell_line</th>\n",
       "      <th>samples_a_single_cell</th>\n",
       "      <th>samples_b_single_cell</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>GSE124939</td>\n",
       "      <td>164</td>\n",
       "      <td>Hypersensitive interferon responses in lupus k...</td>\n",
       "      <td>GEO</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi...</td>\n",
       "      <td>0</td>\n",
       "      <td>80</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    study_id  total_sample_count  \\\n",
       "0  GSE124939                 164   \n",
       "\n",
       "                                         study_title study_type  \\\n",
       "0  Hypersensitive interferon responses in lupus k...        GEO   \n",
       "\n",
       "                                          study_link  samples_a  samples_b  \\\n",
       "0  https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi...          0         80   \n",
       "\n",
       "   samples_a_cell_line  samples_b_cell_line  samples_a_single_cell  \\\n",
       "0                    0                    0                      0   \n",
       "\n",
       "   samples_b_single_cell  \n",
       "0                      0  "
      ]
     },
     "execution_count": 135,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def get_study_info(hgnc_id,ensemble_id,lowrange,highrange,entity,term):\n",
    "    url = \"https://preview.nferx.com/dv/201912/rnaapi2/term_distribution?\"\n",
    "    params = {\n",
    "        \"entity_type\":entity,\n",
    "        \"range_a_min\":lowrange[0],\n",
    "        \"range_a_max\":lowrange[1],\n",
    "        \"range_b_min\":highrange[0],\n",
    "        \"range_b_max\":highrange[1],\n",
    "        \"enriched_group\": \"B\",\n",
    "        \"exclude_zero\": True,\n",
    "        \"token_extraction\": \"ngram\",\n",
    "        \"single_cell_filter\":\"not_single_cell\",\n",
    "        \"cell_line_filter\":\"not_cell_line\",\n",
    "        \"exclude_zero\":True,\n",
    "        \"gene_id\":ensemble_id,\n",
    "        \"hgnc_id\":hgnc_id,\n",
    "        \"term\":term\n",
    "    }\n",
    "    x = requests.get(url, params=params, headers = headers)\n",
    "    required_table = json_normalize(x.json()['result']['term_features']['studies_found'])\n",
    "    return(required_table)\n",
    "all_IL1A_disease_enriched_high = get_study_info(\"il1a\",\"ensg00000115008\",[0,0.06],[3.68,14],\"root-diseases\",\"lupus\")\n",
    "all_IL1A_disease_enriched_high"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>study_id</th>\n",
       "      <th>total_sample_count</th>\n",
       "      <th>study_title</th>\n",
       "      <th>study_type</th>\n",
       "      <th>study_link</th>\n",
       "      <th>samples_a</th>\n",
       "      <th>samples_b</th>\n",
       "      <th>samples_a_cell_line</th>\n",
       "      <th>samples_b_cell_line</th>\n",
       "      <th>samples_a_single_cell</th>\n",
       "      <th>samples_b_single_cell</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>GSE86884</td>\n",
       "      <td>96</td>\n",
       "      <td>Differentially Expressed Gene Transcripts Usin...</td>\n",
       "      <td>GEO</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi...</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>GSE110685</td>\n",
       "      <td>53</td>\n",
       "      <td>Low density granulocytes and their gene signat...</td>\n",
       "      <td>GEO</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi...</td>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>GSE139358</td>\n",
       "      <td>41</td>\n",
       "      <td>Transcriptomic, epigenetic and functional anal...</td>\n",
       "      <td>GEO</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi...</td>\n",
       "      <td>0</td>\n",
       "      <td>15</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    study_id  total_sample_count  \\\n",
       "0   GSE86884                  96   \n",
       "1  GSE110685                  53   \n",
       "2  GSE139358                  41   \n",
       "\n",
       "                                         study_title study_type  \\\n",
       "0  Differentially Expressed Gene Transcripts Usin...        GEO   \n",
       "1  Low density granulocytes and their gene signat...        GEO   \n",
       "2  Transcriptomic, epigenetic and functional anal...        GEO   \n",
       "\n",
       "                                          study_link  samples_a  samples_b  \\\n",
       "0  https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi...          0          2   \n",
       "1  https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi...          0         11   \n",
       "2  https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi...          0         15   \n",
       "\n",
       "   samples_a_cell_line  samples_b_cell_line  samples_a_single_cell  \\\n",
       "0                    0                    0                      0   \n",
       "1                    0                    0                      0   \n",
       "2                    0                    0                      0   \n",
       "\n",
       "   samples_b_single_cell  \n",
       "0                      0  \n",
       "1                      0  \n",
       "2                      0  "
      ]
     },
     "execution_count": 136,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_IL1A_disease_enriched_low = get_study_info(\"il1a\",\"ensg00000115008\",[0,0],[0,0.06],\"root-diseases\",\"lupus\")\n",
    "all_IL1A_disease_enriched_low"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>study_id</th>\n",
       "      <th>total_sample_count</th>\n",
       "      <th>study_title</th>\n",
       "      <th>study_type</th>\n",
       "      <th>study_link</th>\n",
       "      <th>samples_a</th>\n",
       "      <th>samples_b</th>\n",
       "      <th>samples_a_cell_line</th>\n",
       "      <th>samples_b_cell_line</th>\n",
       "      <th>samples_a_single_cell</th>\n",
       "      <th>samples_b_single_cell</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>GSE124939</td>\n",
       "      <td>164</td>\n",
       "      <td>Hypersensitive interferon responses in lupus k...</td>\n",
       "      <td>GEO</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>GSE139358</td>\n",
       "      <td>41</td>\n",
       "      <td>Transcriptomic, epigenetic and functional anal...</td>\n",
       "      <td>GEO</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi...</td>\n",
       "      <td>15</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    study_id  total_sample_count  \\\n",
       "0  GSE124939                 164   \n",
       "1  GSE139358                  41   \n",
       "\n",
       "                                         study_title study_type  \\\n",
       "0  Hypersensitive interferon responses in lupus k...        GEO   \n",
       "1  Transcriptomic, epigenetic and functional anal...        GEO   \n",
       "\n",
       "                                          study_link  samples_a  samples_b  \\\n",
       "0  https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi...          0          1   \n",
       "1  https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi...         15          5   \n",
       "\n",
       "   samples_a_cell_line  samples_b_cell_line  samples_a_single_cell  \\\n",
       "0                    0                    0                      0   \n",
       "1                    0                    0                      0   \n",
       "\n",
       "   samples_b_single_cell  \n",
       "0                      0  \n",
       "1                      0  "
      ]
     },
     "execution_count": 137,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_IL1A_disease_enriched_medium = get_study_info(\"il1a\",\"ensg00000115008\",[0,0.059],[0.06,3.68],\"root-diseases\",\"lupus\")\n",
    "all_IL1A_disease_enriched_medium"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 138,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>study_id</th>\n",
       "      <th>total_sample_count</th>\n",
       "      <th>study_title</th>\n",
       "      <th>study_type</th>\n",
       "      <th>study_link</th>\n",
       "      <th>samples_a</th>\n",
       "      <th>samples_b</th>\n",
       "      <th>samples_a_cell_line</th>\n",
       "      <th>samples_b_cell_line</th>\n",
       "      <th>samples_a_single_cell</th>\n",
       "      <th>samples_b_single_cell</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>GSE124939</td>\n",
       "      <td>164</td>\n",
       "      <td>Hypersensitive interferon responses in lupus k...</td>\n",
       "      <td>GEO</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi...</td>\n",
       "      <td>0</td>\n",
       "      <td>81</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>GSE139358</td>\n",
       "      <td>41</td>\n",
       "      <td>Transcriptomic, epigenetic and functional anal...</td>\n",
       "      <td>GEO</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi...</td>\n",
       "      <td>0</td>\n",
       "      <td>20</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>GSE86884</td>\n",
       "      <td>96</td>\n",
       "      <td>Differentially Expressed Gene Transcripts Usin...</td>\n",
       "      <td>GEO</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi...</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>GSE110685</td>\n",
       "      <td>53</td>\n",
       "      <td>Low density granulocytes and their gene signat...</td>\n",
       "      <td>GEO</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi...</td>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    study_id  total_sample_count  \\\n",
       "0  GSE124939                 164   \n",
       "1  GSE139358                  41   \n",
       "2   GSE86884                  96   \n",
       "3  GSE110685                  53   \n",
       "\n",
       "                                         study_title study_type  \\\n",
       "0  Hypersensitive interferon responses in lupus k...        GEO   \n",
       "1  Transcriptomic, epigenetic and functional anal...        GEO   \n",
       "2  Differentially Expressed Gene Transcripts Usin...        GEO   \n",
       "3  Low density granulocytes and their gene signat...        GEO   \n",
       "\n",
       "                                          study_link  samples_a  samples_b  \\\n",
       "0  https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi...          0         81   \n",
       "1  https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi...          0         20   \n",
       "2  https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi...          0          2   \n",
       "3  https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi...          0         11   \n",
       "\n",
       "   samples_a_cell_line  samples_b_cell_line  samples_a_single_cell  \\\n",
       "0                    0                    0                      0   \n",
       "1                    0                    0                      0   \n",
       "2                    0                    0                      0   \n",
       "3                    0                    0                      0   \n",
       "\n",
       "   samples_b_single_cell  \n",
       "0                      0  \n",
       "1                      0  \n",
       "2                      0  \n",
       "3                      0  "
      ]
     },
     "execution_count": 138,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_IL1A_disease_enriched_total = get_study_info(\"il1a\",\"ensg00000115008\",[0,0.0001],[0.001,14],\"root-diseases\",\"lupus\")\n",
    "all_IL1A_disease_enriched_total"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "4"
      ]
     },
     "execution_count": 140,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_IL1A_disease_enriched_total.co"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 234,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_disease_numbers(diseasename):\n",
    "    all_IL1A_disease_enriched_all = get_study_info(\"il1a\",\"ensg00000115008\",[0,0.0001],[0.001,14],\"root-diseases\",diseasename)\n",
    "    \n",
    "    all_IL1A_disease_enriched_high = get_study_info(\"il1a\",\"ensg00000115008\",[0,0.06],[3.68,14],\"root-diseases\",diseasename)\n",
    "    \n",
    "    all_IL1A_disease_enriched_med = get_study_info(\"il1a\",\"ensg00000115008\",[0,0.059],[0.06,3.68],\"root-diseases\",diseasename)\n",
    "    \n",
    "    try:\n",
    "        all_IL1A_disease_enriched_low = get_study_info(\"il1a\",\"ensg00000115008\",[0,0],[0,0.0599],\"root-diseases\",diseasename)\n",
    "    except TypeError:\n",
    "        all_IL1A_disease_enriched_low = pd.DataFrame(columns = list(all_IL1A_disease_enriched_high.columns))\n",
    "    \n",
    "    \n",
    "    combined_result = str(all_IL1A_disease_enriched_high['samples_b'].sum()) + \",\" + str(all_IL1A_disease_enriched_high.shape[0]) + \",\" + str(all_IL1A_disease_enriched_med['samples_b'].sum()) + \",\" + str(all_IL1A_disease_enriched_med.shape[0])  + \",\" + str(all_IL1A_disease_enriched_low['samples_b'].sum()) + \",\" + str(all_IL1A_disease_enriched_low.shape[0]) + \",\" + str(all_IL1A_disease_enriched_all['samples_b'].sum()) + \",\" + str(all_IL1A_disease_enriched_all.shape[0])\n",
    "    \n",
    "    return(combined_result)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 282,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_disease_numbers2(diseasename):\n",
    "    all_IL1A_disease_enriched_all = get_study_info(\"il1a\",\"ensg00000115008\",[0,0.0001],[0.001,14],\"root-diseases\",diseasename)\n",
    "    \n",
    "    try:\n",
    "        all_IL1A_disease_enriched_high = get_study_info(\"il1a\",\"ensg00000115008\",[0,0.06],[3.68,14],\"root-diseases\",diseasename)\n",
    "    except TypeError:\n",
    "        all_IL1A_disease_enriched_high = pd.DataFrame(columns=list(all_IL1A_disease_enriched_all.columns))\n",
    "    \n",
    "    try:\n",
    "        all_IL1A_disease_enriched_med = get_study_info(\"il1a\",\"ensg00000115008\",[0,0.059],[0.06,3.68],\"root-diseases\",diseasename)\n",
    "    except TypeError:\n",
    "        all_IL1A_disease_enriched_med = pd.DataFrame(columns=list(all_IL1A_disease_enriched_all.columns))\n",
    "    \n",
    "    try:\n",
    "        all_IL1A_disease_enriched_low = get_study_info(\"il1a\",\"ensg00000115008\",[0,0],[0,0.0599],\"root-diseases\",diseasename)\n",
    "    except TypeError:\n",
    "        all_IL1A_disease_enriched_low = pd.DataFrame(columns=list(all_IL1A_disease_enriched_all.columns))\n",
    "    \n",
    "    \n",
    "    combined_result = str(all_IL1A_disease_enriched_high['samples_b'].sum()) + \",\" + str(all_IL1A_disease_enriched_high.shape[0]) + \",\" + str(all_IL1A_disease_enriched_med['samples_b'].sum()) + \",\" + str(all_IL1A_disease_enriched_med.shape[0])  + \",\" + str(all_IL1A_disease_enriched_low['samples_b'].sum()) + \",\" + str(all_IL1A_disease_enriched_low.shape[0]) + \",\" + str(all_IL1A_disease_enriched_all['samples_b'].sum()) + \",\" + str(all_IL1A_disease_enriched_all.shape[0])\n",
    "    \n",
    "    return(combined_result)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 190,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>study_id</th>\n",
       "      <th>total_sample_count</th>\n",
       "      <th>study_title</th>\n",
       "      <th>study_type</th>\n",
       "      <th>study_link</th>\n",
       "      <th>samples_a</th>\n",
       "      <th>samples_b</th>\n",
       "      <th>samples_a_cell_line</th>\n",
       "      <th>samples_b_cell_line</th>\n",
       "      <th>samples_a_single_cell</th>\n",
       "      <th>samples_b_single_cell</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>GSE110685</td>\n",
       "      <td>53</td>\n",
       "      <td>Low density granulocytes and their gene signat...</td>\n",
       "      <td>GEO</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi...</td>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>GSE139358</td>\n",
       "      <td>41</td>\n",
       "      <td>Transcriptomic, epigenetic and functional anal...</td>\n",
       "      <td>GEO</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi...</td>\n",
       "      <td>0</td>\n",
       "      <td>15</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>GSE86884</td>\n",
       "      <td>96</td>\n",
       "      <td>Differentially Expressed Gene Transcripts Usin...</td>\n",
       "      <td>GEO</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi...</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    study_id  total_sample_count  \\\n",
       "0  GSE110685                  53   \n",
       "1  GSE139358                  41   \n",
       "2   GSE86884                  96   \n",
       "\n",
       "                                         study_title study_type  \\\n",
       "0  Low density granulocytes and their gene signat...        GEO   \n",
       "1  Transcriptomic, epigenetic and functional anal...        GEO   \n",
       "2  Differentially Expressed Gene Transcripts Usin...        GEO   \n",
       "\n",
       "                                          study_link  samples_a  samples_b  \\\n",
       "0  https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi...          0         11   \n",
       "1  https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi...          0         15   \n",
       "2  https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi...          0          2   \n",
       "\n",
       "   samples_a_cell_line  samples_b_cell_line  samples_a_single_cell  \\\n",
       "0                    0                    0                      0   \n",
       "1                    0                    0                      0   \n",
       "2                    0                    0                      0   \n",
       "\n",
       "   samples_b_single_cell  \n",
       "0                      0  \n",
       "1                      0  \n",
       "2                      0  "
      ]
     },
     "execution_count": 190,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "get_study_info(\"il1a\",\"ensg00000115008\",[0,0],[0,0.06],\"root-diseases\",\"lupus\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 256,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['study_id', 'total_sample_count', 'study_title', 'study_type',\n",
       "       'study_link', 'samples_a', 'samples_b', 'samples_a_cell_line',\n",
       "       'samples_b_cell_line', 'samples_a_single_cell',\n",
       "       'samples_b_single_cell'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 256,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "get_study_info(\"il1a\",\"ensg00000115008\",[0,0.06],[3.68,14],\"root-diseases\",\"lupus\").columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 210,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>study_id</th>\n",
       "      <th>total_sample_count</th>\n",
       "      <th>study_title</th>\n",
       "      <th>study_type</th>\n",
       "      <th>study_link</th>\n",
       "      <th>samples_a</th>\n",
       "      <th>samples_b</th>\n",
       "      <th>samples_a_cell_line</th>\n",
       "      <th>samples_b_cell_line</th>\n",
       "      <th>samples_a_single_cell</th>\n",
       "      <th>samples_b_single_cell</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>GSE110685</td>\n",
       "      <td>53</td>\n",
       "      <td>Low density granulocytes and their gene signat...</td>\n",
       "      <td>GEO</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi...</td>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>GSE139358</td>\n",
       "      <td>41</td>\n",
       "      <td>Transcriptomic, epigenetic and functional anal...</td>\n",
       "      <td>GEO</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi...</td>\n",
       "      <td>0</td>\n",
       "      <td>15</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>GSE86884</td>\n",
       "      <td>96</td>\n",
       "      <td>Differentially Expressed Gene Transcripts Usin...</td>\n",
       "      <td>GEO</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi...</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    study_id  total_sample_count  \\\n",
       "0  GSE110685                  53   \n",
       "1  GSE139358                  41   \n",
       "2   GSE86884                  96   \n",
       "\n",
       "                                         study_title study_type  \\\n",
       "0  Low density granulocytes and their gene signat...        GEO   \n",
       "1  Transcriptomic, epigenetic and functional anal...        GEO   \n",
       "2  Differentially Expressed Gene Transcripts Usin...        GEO   \n",
       "\n",
       "                                          study_link  samples_a  samples_b  \\\n",
       "0  https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi...          0         11   \n",
       "1  https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi...          0         15   \n",
       "2  https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi...          0          2   \n",
       "\n",
       "   samples_a_cell_line  samples_b_cell_line  samples_a_single_cell  \\\n",
       "0                    0                    0                      0   \n",
       "1                    0                    0                      0   \n",
       "2                    0                    0                      0   \n",
       "\n",
       "   samples_b_single_cell  \n",
       "0                      0  \n",
       "1                      0  \n",
       "2                      0  "
      ]
     },
     "execution_count": 210,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "get_study_info(\"il1a\",\"ensg00000115008\",[0,0],[0,0.059],\"root-diseases\",\"lupus\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 292,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'1,1,21,2,4,1,26,3'"
      ]
     },
     "execution_count": 292,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "get_disease_numbers2(\"tuberculosis\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 262,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'3,1,17,2,0,0,20,2'"
      ]
     },
     "execution_count": 262,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "get_disease_numbers2(\"recessive_dystrophic_epidermolysis_bullosa\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 255,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>study_id</th>\n",
       "      <th>total_sample_count</th>\n",
       "      <th>study_title</th>\n",
       "      <th>study_type</th>\n",
       "      <th>study_link</th>\n",
       "      <th>samples_a</th>\n",
       "      <th>samples_b</th>\n",
       "      <th>samples_a_cell_line</th>\n",
       "      <th>samples_b_cell_line</th>\n",
       "      <th>samples_a_single_cell</th>\n",
       "      <th>samples_b_single_cell</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>GSE111927</td>\n",
       "      <td>11</td>\n",
       "      <td>Effect of chorioamnionitis exposure on neonata...</td>\n",
       "      <td>GEO</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi...</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    study_id  total_sample_count  \\\n",
       "0  GSE111927                  11   \n",
       "\n",
       "                                         study_title study_type  \\\n",
       "0  Effect of chorioamnionitis exposure on neonata...        GEO   \n",
       "\n",
       "                                          study_link  samples_a  samples_b  \\\n",
       "0  https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi...          0          5   \n",
       "\n",
       "   samples_a_cell_line  samples_b_cell_line  samples_a_single_cell  \\\n",
       "0                    0                    0                      0   \n",
       "\n",
       "   samples_b_single_cell  \n",
       "0                      0  "
      ]
     },
     "execution_count": 255,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "get_study_info(\"il1a\",\"ensg00000115008\",[0,0.06],[3.86,14],\"root-diseases\",\"chorioamnionitis\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 220,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell_line_data_a</th>\n",
       "      <th>cell_line_data_b</th>\n",
       "      <th>range_a_ratio</th>\n",
       "      <th>range_b_ratio</th>\n",
       "      <th>range_a_count</th>\n",
       "      <th>range_b_count</th>\n",
       "      <th>range_a_count_cell_line</th>\n",
       "      <th>range_b_count_cell_line</th>\n",
       "      <th>range_a_count_single_cell</th>\n",
       "      <th>range_b_count_single_cell</th>\n",
       "      <th>term</th>\n",
       "      <th>confounding_terms</th>\n",
       "      <th>studies_found</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0</td>\n",
       "      <td>40</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>acute_on_chronic_liver_failure</td>\n",
       "      <td>[{'term': 'liver', 'samples_count': 40, 'all_c...</td>\n",
       "      <td>[{'study_id': 'GSE93265', 'total_sample_count'...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  cell_line_data_a cell_line_data_b  range_a_ratio  range_b_ratio  \\\n",
       "0             None             None              0           0.01   \n",
       "\n",
       "   range_a_count  range_b_count  range_a_count_cell_line  \\\n",
       "0              0             40                        0   \n",
       "\n",
       "   range_b_count_cell_line  range_a_count_single_cell  \\\n",
       "0                        0                          0   \n",
       "\n",
       "   range_b_count_single_cell                            term  \\\n",
       "0                          0  acute_on_chronic_liver_failure   \n",
       "\n",
       "                                   confounding_terms  \\\n",
       "0  [{'term': 'liver', 'samples_count': 40, 'all_c...   \n",
       "\n",
       "                                       studies_found  \n",
       "0  [{'study_id': 'GSE93265', 'total_sample_count'...  "
      ]
     },
     "execution_count": 220,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "url = \"https://preview.nferx.com/dv/201912/rnaapi2/term_distribution?\"\n",
    "\n",
    "#https://preview.nferx.com/rnaapi2/entity_distribution?hgnc_id=il1a&gene_id=ensg00000115008&range_a_min=0&range_a_max=0.07&range_b_min=3.86&range_b_max=14&entity_type=root-cells-and-tissues&enriched_group=B&cell_line_filter=not_cell_line&single_cell_filter=not_single_cell&exclude_zero=true&token_extraction=ngram\n",
    "\n",
    "params = {\n",
    "    \"entity_type\":\"root-diseases\",\n",
    "    \"range_a_min\":0,\n",
    "    \"range_a_max\":0.06,\n",
    "    \"range_b_min\":3.86,\n",
    "    \"range_b_max\":14,\n",
    "    \"enriched_group\": \"B\",\n",
    "    \"exclude_zero\": True,\n",
    "    \"token_extraction\": \"ngram\",\n",
    "    \"single_cell_filter\":\"not_single_cell\",\n",
    "    \"cell_line_filter\":\"all\",\n",
    "    \"exclude_zero\":True,\n",
    "    \"gene_id\":\"ensg00000115008\",\n",
    "    \"hgnc_id\":\"IL1A\",\n",
    "    \"term\" : \"acute_on_chronic_liver_failure\"\n",
    "}\n",
    "\n",
    "x = requests.get(url, params=params, headers = headers)\n",
    "all_data = json_normalize(x.json()['result']['term_features'])\n",
    "all_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 225,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1"
      ]
     },
     "execution_count": 225,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(all_data['studies_found'].values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
